{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "domestic-airport",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Linux/MacOS detected! \n",
      "\n",
      "\n",
      "Checking if samtools is installed ... \n",
      "\n",
      "\n",
      "samtools was located! \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"scPipe\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "chief-cowboy",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.folder <- system.file(\"extdata\", package = \"scPipe\", mustWork = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cathedral-joseph",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata'"
      ],
      "text/latex": [
       "'/SGRNJ01/Public/Software/conda\\_env/scpipe/lib/R/library/scPipe/extdata'"
      ],
      "text/markdown": [
       "'/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata'"
      ],
      "text/plain": [
       "[1] \"/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data.folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "silent-string",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_folder=\"/SGRNJ06/randd/USER/cjj/celedev/atac/20230815ds/test3_10xdata\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "brilliant-draft",
   "metadata": {},
   "outputs": [],
   "source": [
    "?sc_atac_create_sce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "accessory-recorder",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SCE object is saved in:\n",
      "\t/SGRNJ06/randd/USER/cjj/celedev/atac/20230815ds/test3_10xdata/scPipe_atac_SCEobject.rds\n",
      "\n",
      "\n",
      "\n",
      "processing file: rmd_report_skeleton.Rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/59                            \n",
      "2/59 [setup]                    \n",
      "3/59                            \n",
      "4/59 [unnamed-chunk-1]          \n",
      "5/59                            \n",
      "6/59 [inputs]                   \n",
      "7/59                            \n",
      "8/59 [unnamed-chunk-2]          \n",
      "9/59                            \n",
      "10/59 [unnamed-chunk-3]          \n",
      "11/59                            \n",
      "12/59 [unnamed-chunk-4]          \n",
      "13/59                            \n",
      "14/59 [alignment-stats-plot]     \n",
      "15/59                            \n",
      "16/59 [unnamed-chunk-5]          \n",
      "17/59                            \n",
      "18/59 [unnamed-chunk-6]          \n",
      "19/59                            \n",
      "20/59 [barcode-alignment-barplot]\n",
      "21/59                            \n",
      "22/59 [unnamed-chunk-7]          \n",
      "23/59                            \n",
      "24/59 [insert-size-dist]         \n",
      "25/59                            \n",
      "26/59 [unnamed-chunk-8]          \n",
      "27/59                            \n",
      "28/59 [frag-peak]                \n",
      "29/59                            \n",
      "30/59 [frag-mito]                \n",
      "31/59                            \n",
      "32/59 [frag-density]             \n",
      "33/59                            \n",
      "34/59 [barcode-alignment-dist]   \n",
      "35/59                            \n",
      "36/59 [percell-counts]           \n",
      "37/59                            \n",
      "38/59 [percell-features]         \n",
      "39/59                            \n",
      "40/59 [unnamed-chunk-9]          \n",
      "41/59                            \n",
      "42/59 [perfeature-counts]        \n",
      "43/59                            \n",
      "44/59 [perfeature-cells]         \n",
      "45/59                            \n",
      "46/59 [corr-cells]               \n",
      "47/59                            \n",
      "48/59 [corr-features]            \n",
      "49/59                            \n",
      "50/59 [overall-overlap]          \n",
      "51/59                            \n",
      "52/59 [anno-overlap-boxes]       \n",
      "53/59                            \n",
      "54/59 [tss-enrichment]           \n",
      "55/59                            \n",
      "56/59 [unnamed-chunk-10]         \n",
      "57/59                            \n",
      "58/59 [interactive-umap]         \n",
      "59/59                            \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: rmd_report_skeleton.knit.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/SGRNJ01/Public/Software/anaconda3/bin/pandoc +RTS -K512m -RTS rmd_report_skeleton.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /SGRNJ06/randd/USER/cjj/celedev/atac/20230815ds/test3_10xdata/scPipe_atac_stats/scPipe_atac_report.html --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --id-prefix section- --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmpDF3wZn/rmarkdown-str1501f322be0.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Output created: /SGRNJ06/randd/USER/cjj/celedev/atac/20230815ds/test3_10xdata/scPipe_atac_stats/scPipe_atac_report.html\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sce <- sc_atac_create_sce(input_folder = output_folder,\n",
    "                   organism     = \"hg38\",\n",
    "                   feature_type = \"peak\",\n",
    "                   pheno_data   = NULL,\n",
    "                   report       = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "weekly-nothing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "naval-suspect",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "inner-consequence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>sc_aligning {scPipe}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2 id='sc_aligning'>aligning the demultiplexed FASTQ reads using the Rsubread:align()</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>after we run the <code>sc_trim_barcode</code> or <code>sc_atac_trim_barcode</code> to demultiplex the fastq files, we are using this\n",
       "function to align those fastq files to a known reference.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>sc_aligning(\n",
       "  R1,\n",
       "  R2 = NULL,\n",
       "  tech = \"atac\",\n",
       "  index_path = NULL,\n",
       "  ref = NULL,\n",
       "  output_folder = NULL,\n",
       "  output_file = NULL,\n",
       "  input_format = \"FASTQ\",\n",
       "  output_format = \"BAM\",\n",
       "  type = \"dna\",\n",
       "  nthreads = 1\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"sc_aligning_:_R1\">R1</code></td>\n",
       "<td>\n",
       "<p>a mandatory character vector including names of files that include sequence reads to be aligned. For paired-end reads, this gives the list of files including first reads in each library. File format is FASTQ/FASTA by default.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_R2\">R2</code></td>\n",
       "<td>\n",
       "<p>a character vector, the second fastq file, which is required if the data is paired-end</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_tech\">tech</code></td>\n",
       "<td>\n",
       "<p>a character string giving the sequencing technology. Possible value includes &quot;atac&quot; or &quot;rna&quot;</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_index_path\">index_path</code></td>\n",
       "<td>\n",
       "<p>character string specifying the path/basename of the index files, if the Rsubread genome build is available</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_ref\">ref</code></td>\n",
       "<td>\n",
       "<p>a character string specifying the path to reference genome file (.fasta, .fa format)</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_output_folder\">output_folder</code></td>\n",
       "<td>\n",
       "<p>a character string, the name of the output folder</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_output_file\">output_file</code></td>\n",
       "<td>\n",
       "<p>a character vector specifying names of output files. By default, names of output files are set as the file names provided in R1 added with an suffix string</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_input_format\">input_format</code></td>\n",
       "<td>\n",
       "<p>a string indicating the input format</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_output_format\">output_format</code></td>\n",
       "<td>\n",
       "<p>a string indicating the output format</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_type\">type</code></td>\n",
       "<td>\n",
       "<p>type of sequencing data ('RNA' or 'DNA')</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_aligning_:_nthreads\">nthreads</code></td>\n",
       "<td>\n",
       "<p>numeric value giving the number of threads used for mapping.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>the file path of the output aligned BAM file\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>## Not run: \n",
       "sc_aligning(index_path,\n",
       "    tech = 'atac',\n",
       "    R1,  \n",
       "    R2, \n",
       "    nthreads  = 6) \n",
       "\n",
       "## End(Not run)\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>scPipe</em> version 2.0.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{sc\\_aligning}{aligning the demultiplexed FASTQ reads using the Rsubread:align()}{sc.Rul.aligning}\n",
       "%\n",
       "\\begin{Description}\n",
       "after we run the \\code{sc\\_trim\\_barcode} or \\code{sc\\_atac\\_trim\\_barcode} to demultiplex the fastq files, we are using this\n",
       "function to align those fastq files to a known reference.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "sc_aligning(\n",
       "  R1,\n",
       "  R2 = NULL,\n",
       "  tech = \"atac\",\n",
       "  index_path = NULL,\n",
       "  ref = NULL,\n",
       "  output_folder = NULL,\n",
       "  output_file = NULL,\n",
       "  input_format = \"FASTQ\",\n",
       "  output_format = \"BAM\",\n",
       "  type = \"dna\",\n",
       "  nthreads = 1\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{R1}] a mandatory character vector including names of files that include sequence reads to be aligned. For paired-end reads, this gives the list of files including first reads in each library. File format is FASTQ/FASTA by default.\n",
       "\n",
       "\\item[\\code{R2}] a character vector, the second fastq file, which is required if the data is paired-end\n",
       "\n",
       "\\item[\\code{tech}] a character string giving the sequencing technology. Possible value includes \"atac\" or \"rna\"\n",
       "\n",
       "\\item[\\code{index\\_path}] character string specifying the path/basename of the index files, if the Rsubread genome build is available\n",
       "\n",
       "\\item[\\code{ref}] a character string specifying the path to reference genome file (.fasta, .fa format)\n",
       "\n",
       "\\item[\\code{output\\_folder}] a character string, the name of the output folder\n",
       "\n",
       "\\item[\\code{output\\_file}] a character vector specifying names of output files. By default, names of output files are set as the file names provided in R1 added with an suffix string\n",
       "\n",
       "\\item[\\code{input\\_format}] a string indicating the input format\n",
       "\n",
       "\\item[\\code{output\\_format}] a string indicating the output format\n",
       "\n",
       "\\item[\\code{type}] type of sequencing data (`RNA` or `DNA`)\n",
       "\n",
       "\\item[\\code{nthreads}] numeric value giving the number of threads used for mapping.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "the file path of the output aligned BAM file\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "## Not run: \n",
       "sc_aligning(index_path,\n",
       "    tech = 'atac',\n",
       "    R1,  \n",
       "    R2, \n",
       "    nthreads  = 6) \n",
       "\n",
       "## End(Not run)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "sc_aligning               package:scPipe               R Documentation\n",
       "\n",
       "_\ba_\bl_\bi_\bg_\bn_\bi_\bn_\bg _\bt_\bh_\be _\bd_\be_\bm_\bu_\bl_\bt_\bi_\bp_\bl_\be_\bx_\be_\bd _\bF_\bA_\bS_\bT_\bQ _\br_\be_\ba_\bd_\bs _\bu_\bs_\bi_\bn_\bg _\bt_\bh_\be _\bR_\bs_\bu_\bb_\br_\be_\ba_\bd:_\ba_\bl_\bi_\bg_\bn()\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     after we run the ‘sc_trim_barcode’ or ‘sc_atac_trim_barcode’ to\n",
       "     demultiplex the fastq files, we are using this function to align\n",
       "     those fastq files to a known reference.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     sc_aligning(\n",
       "       R1,\n",
       "       R2 = NULL,\n",
       "       tech = \"atac\",\n",
       "       index_path = NULL,\n",
       "       ref = NULL,\n",
       "       output_folder = NULL,\n",
       "       output_file = NULL,\n",
       "       input_format = \"FASTQ\",\n",
       "       output_format = \"BAM\",\n",
       "       type = \"dna\",\n",
       "       nthreads = 1\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "      R1: a mandatory character vector including names of files that\n",
       "          include sequence reads to be aligned. For paired-end reads,\n",
       "          this gives the list of files including first reads in each\n",
       "          library. File format is FASTQ/FASTA by default.\n",
       "\n",
       "      R2: a character vector, the second fastq file, which is required\n",
       "          if the data is paired-end\n",
       "\n",
       "    tech: a character string giving the sequencing technology. Possible\n",
       "          value includes \"atac\" or \"rna\"\n",
       "\n",
       "index_path: character string specifying the path/basename of the index\n",
       "          files, if the Rsubread genome build is available\n",
       "\n",
       "     ref: a character string specifying the path to reference genome\n",
       "          file (.fasta, .fa format)\n",
       "\n",
       "output_folder: a character string, the name of the output folder\n",
       "\n",
       "output_file: a character vector specifying names of output files. By\n",
       "          default, names of output files are set as the file names\n",
       "          provided in R1 added with an suffix string\n",
       "\n",
       "input_format: a string indicating the input format\n",
       "\n",
       "output_format: a string indicating the output format\n",
       "\n",
       "    type: type of sequencing data (`RNA` or `DNA`)\n",
       "\n",
       "nthreads: numeric value giving the number of threads used for mapping.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     the file path of the output aligned BAM file\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     ## Not run:\n",
       "     \n",
       "     sc_aligning(index_path,\n",
       "         tech = 'atac',\n",
       "         R1,  \n",
       "         R2, \n",
       "         nthreads  = 6) \n",
       "     ## End(Not run)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?sc_aligning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "arctic-baltimore",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scpipe",
   "language": "R",
   "name": "scpipe"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
