{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "pressing-engineer",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"scPipe\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moving-method",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reference = \"/SGRNJ06/randd/USER/cjj/cr-atac/refdata-cellranger-arc-GRCh38-2020-A-2.0.0/fasta/genome.fa\"\n",
    "reference = '/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata/small_chr21.fa'\n",
    "r1 = \"/SGRNJ06/randd/USER/cjj/celedev/atac/20230607hs_mu/scPipeATACVignette/demux_completematch_small_chr21_R1.fastq.gz\"\n",
    "r2 = \"/SGRNJ06/randd/USER/cjj/celedev/atac/20230607hs_mu/scPipeATACVignette/demux_completematch_small_chr21_R3.fastq.gz\"\n",
    "output_folder = \"/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/02.atac\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "verbal-yellow",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving the output at location: \n",
      "/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No valid_barcode_file provided; no barcode error correction will occur.\n",
      "Total reads: 35000\n",
      "Total reads removed due to N's in barcodes: 0\n",
      "Total reads removed due to low quality barcodes: 0\n",
      "Total barcodes provided in FASTQ file: 16567\n",
      "time elapsed: 535 milliseconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "data.folder <- system.file(\"extdata\", package = \"scPipe\", mustWork = TRUE)\n",
    "output_folder <- \"/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\"\n",
    "# data fastq files\n",
    "r1      <- file.path(data.folder, \"small_chr21_R1.fastq.gz\") \n",
    "r2      <- file.path(data.folder, \"small_chr21_R3.fastq.gz\") \n",
    "\n",
    "# barcodes in fastq format\n",
    "barcode_fastq      <- file.path(data.folder, \"small_chr21_R2.fastq.gz\") \n",
    "\n",
    "# barcodes in .csv format\n",
    "barcode_1000       <- file.path(data.folder, \"chr21_modified_barcode_1000.csv\")\n",
    "\n",
    "sc_atac_trim_barcode (r1            = r1, \n",
    "                      r2            = r2, \n",
    "                      bc_file       = barcode_fastq, \n",
    "                      rmN           = FALSE,\n",
    "                      rmlow         = FALSE,\n",
    "                      output_folder = output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "particular-potential",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>sc_atac_trim_barcode {scPipe}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2 id='sc_atac_trim_barcode'>demultiplex raw single-cell ATAC-Seq fastq reads</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>single-cell data need to be demultiplexed in order to retain the information of the cell barcodes\n",
       "the data belong to. Here we reformat fastq files so barcode/s (and if available the UMI sequences) are moved from\n",
       "the sequence into the read name. Since scATAC-Seq data are mostly paired-end, both 'r1' and 'r2' are demultiplexed in this function.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>sc_atac_trim_barcode(\n",
       "  r1,\n",
       "  r2,\n",
       "  bc_file = NULL,\n",
       "  valid_barcode_file = \"\",\n",
       "  output_folder = \"\",\n",
       "  umi_start = 0,\n",
       "  umi_length = 0,\n",
       "  umi_in = \"both\",\n",
       "  rmN = FALSE,\n",
       "  rmlow = FALSE,\n",
       "  min_qual = 20,\n",
       "  num_below_min = 2,\n",
       "  id1_st = -0,\n",
       "  id1_len = 16,\n",
       "  id2_st = 0,\n",
       "  id2_len = 16,\n",
       "  no_reverse_complement = FALSE\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_r1\">r1</code></td>\n",
       "<td>\n",
       "<p>read one for pair-end reads.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_r2\">r2</code></td>\n",
       "<td>\n",
       "<p>read two for pair-end reads, NULL if single read.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_bc_file\">bc_file</code></td>\n",
       "<td>\n",
       "<p>the barcode information, can be either in a <code>fastq</code> format (e.g. from 10x-ATAC) or\n",
       "from a <code>.csv</code> file (here the barcode is expected to be on the second column). \n",
       "Currently, for the fastq approach, this can be a list of barcode files.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_valid_barcode_file\">valid_barcode_file</code></td>\n",
       "<td>\n",
       "<p>optional file path of the valid (expected) barcode sequences to be found in the bc_file (.txt, can be txt.gz). \n",
       "Must contain one barcode per line on the second column separated by a comma (default =&quot;&quot;).\n",
       "If given, each barcode from bc_file is matched against the barcode of\n",
       "best fit (allowing a hamming distance of 1). If a FASTQ <code>bc_file</code> is provided, barcodes with a higher mapping quality, as given by\n",
       "the fastq reads quality score are prioritised.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_output_folder\">output_folder</code></td>\n",
       "<td>\n",
       "<p>the output dir for the demultiplexed fastq file, which will contain \n",
       "fastq files with reformatted barcode and UMI into the read name. \n",
       "Files ending in <code>.gz</code> will be automatically compressed.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_umi_start\">umi_start</code></td>\n",
       "<td>\n",
       "<p>if available, the start position of the molecular identifier.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_umi_length\">umi_length</code></td>\n",
       "<td>\n",
       "<p>if available, the start position of the molecular identifier.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_umi_in\">umi_in</code></td>\n",
       "<td>\n",
       "<p>umi_in</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_rmN\">rmN</code></td>\n",
       "<td>\n",
       "<p>logical, whether to remove reads that contains N in UMI or cell barcode.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_rmlow\">rmlow</code></td>\n",
       "<td>\n",
       "<p>logical, whether to remove reads that have low quality barcode sequences</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_min_qual\">min_qual</code></td>\n",
       "<td>\n",
       "<p>the minimum base pair quality that is allowed (default = 20).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_num_below_min\">num_below_min</code></td>\n",
       "<td>\n",
       "<p>the maximum number of base pairs below the quality threshold.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_id1_st\">id1_st</code></td>\n",
       "<td>\n",
       "<p>barcode start position (0-indexed) for read 1, which is an extra parameter that is needed if the\n",
       "<code>bc_file</code> is in a <code>.csv</code> format.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_id1_len\">id1_len</code></td>\n",
       "<td>\n",
       "<p>barcode length for read 1, which is an extra parameter that is needed if the\n",
       "<code>bc_file</code> is in a <code>.csv</code> format.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_id2_st\">id2_st</code></td>\n",
       "<td>\n",
       "<p>barcode start position (0-indexed) for read 2, which is an extra parameter that is needed if the\n",
       "<code>bc_file</code> is in a <code>.csv</code> format.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_id2_len\">id2_len</code></td>\n",
       "<td>\n",
       "<p>barcode length for read 2, which is an extra parameter that is needed if the\n",
       "<code>bc_file</code> is in a <code>.csv</code> format.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"sc_atac_trim_barcode_:_no_reverse_complement\">no_reverse_complement</code></td>\n",
       "<td>\n",
       "<p>specifies if the reverse complement of the barcode sequence should be \n",
       "used for barcode error correction (only when barcode sequences are provided as fastq files). FALSE (default)\n",
       "lets the function decide whether to use reverse complement, and TRUE forces the function to\n",
       "use the forward barcode sequences.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>None (invisible 'NULL')\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>data.folder &lt;- system.file(\"extdata\", package = \"scPipe\", mustWork = TRUE)\n",
       "r1      &lt;- file.path(data.folder, \"small_chr21_R1.fastq.gz\") \n",
       "r2      &lt;- file.path(data.folder, \"small_chr21_R3.fastq.gz\") \n",
       "\n",
       "# Using a barcode fastq file:\n",
       "\n",
       "# barcodes in fastq format\n",
       "barcode_fastq      &lt;- file.path(data.folder, \"small_chr21_R2.fastq.gz\") \n",
       "\n",
       "sc_atac_trim_barcode (\n",
       "r1            = r1, \n",
       "r2            = r2, \n",
       "bc_file       = barcode_fastq,\n",
       "rmN           = TRUE,\n",
       "rmlow         = TRUE,\n",
       "output_folder = tempdir())\n",
       "\n",
       "# Using a barcode csv file:\n",
       "\n",
       "# barcodes in .csv format\n",
       "barcode_1000       &lt;- file.path(data.folder, \"chr21_modified_barcode_1000.csv\")\n",
       "\n",
       "## Not run: \n",
       "sc_atac_trim_barcode (\n",
       "r1            = r1, \n",
       "r2            = r2, \n",
       "bc_file       = barcode_1000, \n",
       "id1_st        = 0,\n",
       "rmN           = TRUE,\n",
       "rmlow         = TRUE,\n",
       "output_folder = tempdir())\n",
       "\n",
       "## End(Not run)\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>scPipe</em> version 2.0.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{sc\\_atac\\_trim\\_barcode}{demultiplex raw single-cell ATAC-Seq fastq reads}{sc.Rul.atac.Rul.trim.Rul.barcode}\n",
       "%\n",
       "\\begin{Description}\n",
       "single-cell data need to be demultiplexed in order to retain the information of the cell barcodes\n",
       "the data belong to. Here we reformat fastq files so barcode/s (and if available the UMI sequences) are moved from\n",
       "the sequence into the read name. Since scATAC-Seq data are mostly paired-end, both `r1` and `r2` are demultiplexed in this function.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "sc_atac_trim_barcode(\n",
       "  r1,\n",
       "  r2,\n",
       "  bc_file = NULL,\n",
       "  valid_barcode_file = \"\",\n",
       "  output_folder = \"\",\n",
       "  umi_start = 0,\n",
       "  umi_length = 0,\n",
       "  umi_in = \"both\",\n",
       "  rmN = FALSE,\n",
       "  rmlow = FALSE,\n",
       "  min_qual = 20,\n",
       "  num_below_min = 2,\n",
       "  id1_st = -0,\n",
       "  id1_len = 16,\n",
       "  id2_st = 0,\n",
       "  id2_len = 16,\n",
       "  no_reverse_complement = FALSE\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{r1}] read one for pair-end reads.\n",
       "\n",
       "\\item[\\code{r2}] read two for pair-end reads, NULL if single read.\n",
       "\n",
       "\\item[\\code{bc\\_file}] the barcode information, can be either in a \\code{fastq} format (e.g. from 10x-ATAC) or\n",
       "from a \\code{.csv} file (here the barcode is expected to be on the second column). \n",
       "Currently, for the fastq approach, this can be a list of barcode files.\n",
       "\n",
       "\\item[\\code{valid\\_barcode\\_file}] optional file path of the valid (expected) barcode sequences to be found in the bc\\_file (.txt, can be txt.gz). \n",
       "Must contain one barcode per line on the second column separated by a comma (default =\"\").\n",
       "If given, each barcode from bc\\_file is matched against the barcode of\n",
       "best fit (allowing a hamming distance of 1). If a FASTQ \\code{bc\\_file} is provided, barcodes with a higher mapping quality, as given by\n",
       "the fastq reads quality score are prioritised.\n",
       "\n",
       "\\item[\\code{output\\_folder}] the output dir for the demultiplexed fastq file, which will contain \n",
       "fastq files with reformatted barcode and UMI into the read name. \n",
       "Files ending in \\code{.gz} will be automatically compressed.\n",
       "\n",
       "\\item[\\code{umi\\_start}] if available, the start position of the molecular identifier.\n",
       "\n",
       "\\item[\\code{umi\\_length}] if available, the start position of the molecular identifier.\n",
       "\n",
       "\\item[\\code{umi\\_in}] umi\\_in\n",
       "\n",
       "\\item[\\code{rmN}] logical, whether to remove reads that contains N in UMI or cell barcode.\n",
       "\n",
       "\\item[\\code{rmlow}] logical, whether to remove reads that have low quality barcode sequences\n",
       "\n",
       "\\item[\\code{min\\_qual}] the minimum base pair quality that is allowed (default = 20).\n",
       "\n",
       "\\item[\\code{num\\_below\\_min}] the maximum number of base pairs below the quality threshold.\n",
       "\n",
       "\\item[\\code{id1\\_st}] barcode start position (0-indexed) for read 1, which is an extra parameter that is needed if the\n",
       "\\code{bc\\_file} is in a \\code{.csv} format.\n",
       "\n",
       "\\item[\\code{id1\\_len}] barcode length for read 1, which is an extra parameter that is needed if the\n",
       "\\code{bc\\_file} is in a \\code{.csv} format.\n",
       "\n",
       "\\item[\\code{id2\\_st}] barcode start position (0-indexed) for read 2, which is an extra parameter that is needed if the\n",
       "\\code{bc\\_file} is in a \\code{.csv} format.\n",
       "\n",
       "\\item[\\code{id2\\_len}] barcode length for read 2, which is an extra parameter that is needed if the\n",
       "\\code{bc\\_file} is in a \\code{.csv} format.\n",
       "\n",
       "\\item[\\code{no\\_reverse\\_complement}] specifies if the reverse complement of the barcode sequence should be \n",
       "used for barcode error correction (only when barcode sequences are provided as fastq files). FALSE (default)\n",
       "lets the function decide whether to use reverse complement, and TRUE forces the function to\n",
       "use the forward barcode sequences.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "None (invisible `NULL`)\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "data.folder <- system.file(\"extdata\", package = \"scPipe\", mustWork = TRUE)\n",
       "r1      <- file.path(data.folder, \"small_chr21_R1.fastq.gz\") \n",
       "r2      <- file.path(data.folder, \"small_chr21_R3.fastq.gz\") \n",
       "\n",
       "# Using a barcode fastq file:\n",
       "\n",
       "# barcodes in fastq format\n",
       "barcode_fastq      <- file.path(data.folder, \"small_chr21_R2.fastq.gz\") \n",
       "\n",
       "sc_atac_trim_barcode (\n",
       "r1            = r1, \n",
       "r2            = r2, \n",
       "bc_file       = barcode_fastq,\n",
       "rmN           = TRUE,\n",
       "rmlow         = TRUE,\n",
       "output_folder = tempdir())\n",
       "\n",
       "# Using a barcode csv file:\n",
       "\n",
       "# barcodes in .csv format\n",
       "barcode_1000       <- file.path(data.folder, \"chr21_modified_barcode_1000.csv\")\n",
       "\n",
       "## Not run: \n",
       "sc_atac_trim_barcode (\n",
       "r1            = r1, \n",
       "r2            = r2, \n",
       "bc_file       = barcode_1000, \n",
       "id1_st        = 0,\n",
       "rmN           = TRUE,\n",
       "rmlow         = TRUE,\n",
       "output_folder = tempdir())\n",
       "\n",
       "## End(Not run)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "sc_atac_trim_barcode          package:scPipe           R Documentation\n",
       "\n",
       "_\bd_\be_\bm_\bu_\bl_\bt_\bi_\bp_\bl_\be_\bx _\br_\ba_\bw _\bs_\bi_\bn_\bg_\bl_\be-_\bc_\be_\bl_\bl _\bA_\bT_\bA_\bC-_\bS_\be_\bq _\bf_\ba_\bs_\bt_\bq _\br_\be_\ba_\bd_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     single-cell data need to be demultiplexed in order to retain the\n",
       "     information of the cell barcodes the data belong to. Here we\n",
       "     reformat fastq files so barcode/s (and if available the UMI\n",
       "     sequences) are moved from the sequence into the read name. Since\n",
       "     scATAC-Seq data are mostly paired-end, both `r1` and `r2` are\n",
       "     demultiplexed in this function.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     sc_atac_trim_barcode(\n",
       "       r1,\n",
       "       r2,\n",
       "       bc_file = NULL,\n",
       "       valid_barcode_file = \"\",\n",
       "       output_folder = \"\",\n",
       "       umi_start = 0,\n",
       "       umi_length = 0,\n",
       "       umi_in = \"both\",\n",
       "       rmN = FALSE,\n",
       "       rmlow = FALSE,\n",
       "       min_qual = 20,\n",
       "       num_below_min = 2,\n",
       "       id1_st = -0,\n",
       "       id1_len = 16,\n",
       "       id2_st = 0,\n",
       "       id2_len = 16,\n",
       "       no_reverse_complement = FALSE\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "      r1: read one for pair-end reads.\n",
       "\n",
       "      r2: read two for pair-end reads, NULL if single read.\n",
       "\n",
       " bc_file: the barcode information, can be either in a ‘fastq’ format\n",
       "          (e.g. from 10x-ATAC) or from a ‘.csv’ file (here the barcode\n",
       "          is expected to be on the second column).  Currently, for the\n",
       "          fastq approach, this can be a list of barcode files.\n",
       "\n",
       "valid_barcode_file: optional file path of the valid (expected) barcode\n",
       "          sequences to be found in the bc_file (.txt, can be txt.gz).\n",
       "          Must contain one barcode per line on the second column\n",
       "          separated by a comma (default =\"\"). If given, each barcode\n",
       "          from bc_file is matched against the barcode of best fit\n",
       "          (allowing a hamming distance of 1). If a FASTQ ‘bc_file’ is\n",
       "          provided, barcodes with a higher mapping quality, as given by\n",
       "          the fastq reads quality score are prioritised.\n",
       "\n",
       "output_folder: the output dir for the demultiplexed fastq file, which\n",
       "          will contain fastq files with reformatted barcode and UMI\n",
       "          into the read name.  Files ending in ‘.gz’ will be\n",
       "          automatically compressed.\n",
       "\n",
       "umi_start: if available, the start position of the molecular\n",
       "          identifier.\n",
       "\n",
       "umi_length: if available, the start position of the molecular\n",
       "          identifier.\n",
       "\n",
       "  umi_in: umi_in\n",
       "\n",
       "     rmN: logical, whether to remove reads that contains N in UMI or\n",
       "          cell barcode.\n",
       "\n",
       "   rmlow: logical, whether to remove reads that have low quality\n",
       "          barcode sequences\n",
       "\n",
       "min_qual: the minimum base pair quality that is allowed (default = 20).\n",
       "\n",
       "num_below_min: the maximum number of base pairs below the quality\n",
       "          threshold.\n",
       "\n",
       "  id1_st: barcode start position (0-indexed) for read 1, which is an\n",
       "          extra parameter that is needed if the ‘bc_file’ is in a\n",
       "          ‘.csv’ format.\n",
       "\n",
       " id1_len: barcode length for read 1, which is an extra parameter that\n",
       "          is needed if the ‘bc_file’ is in a ‘.csv’ format.\n",
       "\n",
       "  id2_st: barcode start position (0-indexed) for read 2, which is an\n",
       "          extra parameter that is needed if the ‘bc_file’ is in a\n",
       "          ‘.csv’ format.\n",
       "\n",
       " id2_len: barcode length for read 2, which is an extra parameter that\n",
       "          is needed if the ‘bc_file’ is in a ‘.csv’ format.\n",
       "\n",
       "no_reverse_complement: specifies if the reverse complement of the\n",
       "          barcode sequence should be used for barcode error correction\n",
       "          (only when barcode sequences are provided as fastq files).\n",
       "          FALSE (default) lets the function decide whether to use\n",
       "          reverse complement, and TRUE forces the function to use the\n",
       "          forward barcode sequences.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     None (invisible `NULL`)\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     data.folder <- system.file(\"extdata\", package = \"scPipe\", mustWork = TRUE)\n",
       "     r1      <- file.path(data.folder, \"small_chr21_R1.fastq.gz\") \n",
       "     r2      <- file.path(data.folder, \"small_chr21_R3.fastq.gz\") \n",
       "     \n",
       "     # Using a barcode fastq file:\n",
       "     \n",
       "     # barcodes in fastq format\n",
       "     barcode_fastq      <- file.path(data.folder, \"small_chr21_R2.fastq.gz\") \n",
       "     \n",
       "     sc_atac_trim_barcode (\n",
       "     r1            = r1, \n",
       "     r2            = r2, \n",
       "     bc_file       = barcode_fastq,\n",
       "     rmN           = TRUE,\n",
       "     rmlow         = TRUE,\n",
       "     output_folder = tempdir())\n",
       "     \n",
       "     # Using a barcode csv file:\n",
       "     \n",
       "     # barcodes in .csv format\n",
       "     barcode_1000       <- file.path(data.folder, \"chr21_modified_barcode_1000.csv\")\n",
       "     \n",
       "     ## Not run:\n",
       "     \n",
       "     sc_atac_trim_barcode (\n",
       "     r1            = r1, \n",
       "     r2            = r2, \n",
       "     bc_file       = barcode_1000, \n",
       "     id1_st        = 0,\n",
       "     rmN           = TRUE,\n",
       "     rmlow         = TRUE,\n",
       "     output_folder = tempdir())\n",
       "     ## End(Not run)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?sc_atac_trim_barcode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "durable-guitar",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ATAC-Seq mode is selected...\n",
      "\n",
      "Genome index location not specified. Looking for the index in/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n",
      "Genome index not found. Creating one in /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out ...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ==========     _____ _    _ ____  _____  ______          _____  \n",
      "        =====         / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "          =====      | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "            ====      \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "              ====    ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "        ==========   |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\n",
      "       Rsubread 2.14.2\n",
      "\n",
      "//================================= setting ==================================\\\\\n",
      "||                                                                            ||\n",
      "||                Index name : genome_index                                   ||\n",
      "||               Index space : base space                                     ||\n",
      "||               Index split : no-split                                       ||\n",
      "||          Repeat threshold : 100 repeats                                    ||\n",
      "||              Gapped index : no                                             ||\n",
      "||                                                                            ||\n",
      "||       Free / total memory : 38.9GB / 688.8GB                               ||\n",
      "||                                                                            ||\n",
      "||               Input files : 1 file in total                                ||\n",
      "||                             o small_chr21.fa                               ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//================================= Running ==================================\\\\\n",
      "||                                                                            ||\n",
      "|| Check the integrity of provided reference sequences ...                    ||\n",
      "|| No format issues were found                                                ||\n",
      "|| Scan uninformative subreads in reference sequences ...                     ||\n",
      "|| 3 uninformative subreads were found.                                       ||\n",
      "|| These subreads were excluded from index building.                          ||\n",
      "|| Estimate the index size...                                                 ||\n",
      "||    8%,   0 mins elapsed, rate=380.7k bps/s                                 ||\n",
      "||   16%,   0 mins elapsed, rate=706.3k bps/s                                 ||\n",
      "||   24%,   0 mins elapsed, rate=1018.7k bps/s                                ||\n",
      "||   33%,   0 mins elapsed, rate=1273.1k bps/s                                ||\n",
      "||   41%,   0 mins elapsed, rate=1537.0k bps/s                                ||\n",
      "||   49%,   0 mins elapsed, rate=1737.9k bps/s                                ||\n",
      "||   58%,   0 mins elapsed, rate=1870.7k bps/s                                ||\n",
      "||   66%,   0 mins elapsed, rate=2062.2k bps/s                                ||\n",
      "||   74%,   0 mins elapsed, rate=2177.5k bps/s                                ||\n",
      "||   83%,   0 mins elapsed, rate=2344.2k bps/s                                ||\n",
      "||   91%,   0 mins elapsed, rate=2549.3k bps/s                                ||\n",
      "|| 3.0 GB of memory is needed for index building.                             ||\n",
      "|| Build the index...                                                         ||\n",
      "||    8%,   0 mins elapsed, rate=35.5k bps/s                                  ||\n",
      "||   16%,   0 mins elapsed, rate=69.6k bps/s                                  ||\n",
      "||   24%,   0 mins elapsed, rate=103.3k bps/s                                 ||\n",
      "||   33%,   0 mins elapsed, rate=135.1k bps/s                                 ||\n",
      "||   41%,   0 mins elapsed, rate=167.2k bps/s                                 ||\n",
      "||   49%,   0 mins elapsed, rate=196.8k bps/s                                 ||\n",
      "||   58%,   0 mins elapsed, rate=225.4k bps/s                                 ||\n",
      "||   66%,   0 mins elapsed, rate=255.1k bps/s                                 ||\n",
      "||   74%,   0 mins elapsed, rate=281.8k bps/s                                 ||\n",
      "||   83%,   0 mins elapsed, rate=310.2k bps/s                                 ||\n",
      "||   91%,   0 mins elapsed, rate=340.1k bps/s                                 ||\n",
      "|| Save current index block...                                                ||\n",
      "||  [ 0.0% finished ]                                                         ||\n",
      "||  [ 10.0% finished ]                                                        ||\n",
      "||  [ 20.0% finished ]                                                        ||\n",
      "||  [ 30.0% finished ]                                                        ||\n",
      "||  [ 40.0% finished ]                                                        ||\n",
      "||  [ 50.0% finished ]                                                        ||\n",
      "||  [ 60.0% finished ]                                                        ||\n",
      "||  [ 70.0% finished ]                                                        ||\n",
      "||  [ 80.0% finished ]                                                        ||\n",
      "||  [ 90.0% finished ]                                                        ||\n",
      "||  [ 100.0% finished ]                                                       ||\n",
      "||                                                                            ||\n",
      "||                      Total running time: 0.2 minutes.                      ||\n",
      "||Index /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x ... ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No partial matches, checking for reads with non-matched barcodes.\n",
      "\n",
      "No reads found with non-matching barcodes.\n",
      "\n",
      "Outputted demultiplexing stats file to/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats//demultiplexing_stats.csv\n",
      "\n",
      "\n",
      "Output file name is not provided. Aligned reads are saved in /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/demux_completematch_small_chr21_R1_aligned.bam\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ==========     _____ _    _ ____  _____  ______          _____  \n",
      "        =====         / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "          =====      | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "            ====      \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "              ====    ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "        ==========   |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\n",
      "       Rsubread 2.14.2\n",
      "\n",
      "//================================= setting ==================================\\\\\n",
      "||                                                                            ||\n",
      "|| Function      : Read alignment (DNA-Seq)                                   ||\n",
      "|| Input file 1  : demux_completematch_small_chr21_R1.fastq.gz                ||\n",
      "|| Input file 2  : demux_completematch_small_chr21_R3.fastq.gz                ||\n",
      "|| Output file   : demux_completematch_small_chr21_R1_aligned.bam (BAM),  ... ||\n",
      "|| Index name    : genome_index                                               ||\n",
      "||                                                                            ||\n",
      "||                    ------------------------------------                    ||\n",
      "||                                                                            ||\n",
      "||                               Threads : 12                                 ||\n",
      "||                          Phred offset : 33                                 ||\n",
      "||               # of extracted subreads : 10                                 ||\n",
      "||                        Min read1 vote : 3                                  ||\n",
      "||                        Min read2 vote : 1                                  ||\n",
      "||                     Max fragment size : 600                                ||\n",
      "||                     Min fragment size : 50                                 ||\n",
      "||                        Max mismatches : 3                                  ||\n",
      "||                      Max indel length : 5                                  ||\n",
      "||            Report multi-mapping reads : yes                                ||\n",
      "|| Max alignments per multi-mapping read : 1                                  ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//================ Running (06-Sep-2023 11:08:05, pid=22357) =================\\\\\n",
      "||                                                                            ||\n",
      "|| Check the input reads.                                                     ||\n",
      "|| The input file contains base space reads.                                  ||\n",
      "|| Initialise the memory objects.                                             ||\n",
      "|| Estimate the mean read length.                                             ||\n",
      "|| The range of Phred scores observed in the data is [2,37]                   ||\n",
      "|| Create the output BAM file.                                                ||\n",
      "|| Check the index.                                                           ||\n",
      "|| Init the voting space.                                                     ||\n",
      "|| Global environment is initialised.                                         ||\n",
      "|| Load the 1-th index block...                                               ||\n",
      "|| The index block has been loaded.                                           ||\n",
      "|| Start read mapping in chunk.                                               ||\n",
      "||   Estimated fragment length : 123 bp                                       ||\n",
      "||                                                                            ||\n",
      "||                           Completed successfully.                          ||\n",
      "||                                                                            ||\n",
      "\\\\====================================    ====================================//\n",
      "\n",
      "//================================   Summary =================================\\\\\n",
      "||                                                                            ||\n",
      "||             Total fragments : 35,000                                       ||\n",
      "||                      Mapped : 18,851 (53.9%)                               ||\n",
      "||             Uniquely mapped : 18,057                                       ||\n",
      "||               Multi-mapping : 794                                          ||\n",
      "||                                                                            ||\n",
      "||                    Unmapped : 16,149                                       ||\n",
      "||                                                                            ||\n",
      "||             Properly paired : 14,524                                       ||\n",
      "||         Not properly paired : 4,327                                        ||\n",
      "||                   Singleton : 1,458                                        ||\n",
      "||                    Chimeric : 0                                            ||\n",
      "||       Unexpected strandness : 114                                          ||\n",
      "||  Unexpected fragment length : 180                                          ||\n",
      "||       Unexpected read order : 2,575                                        ||\n",
      "||                                                                            ||\n",
      "||                      Indels : 197                                          ||\n",
      "||                                                                            ||\n",
      "||                Running time : 0.6 minutes                                  ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "demux_r1        <- file.path(output_folder, \"demux_completematch_small_chr21_R1.fastq.gz\")\n",
    "demux_r2        <- file.path(output_folder, \"demux_completematch_small_chr21_R3.fastq.gz\")\n",
    "reference <- \"/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata/small_chr21.fa\"\n",
    "#output_folder <- \"/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\"\n",
    "#output_folder <- \"/SGRNJ06/randd/USER/cjj/celedev/atac/20230607hs_mu/scPipeATACVignette\"\n",
    "\n",
    "aligned_bam <- sc_aligning(ref = reference, \n",
    "                R1 = demux_r1, \n",
    "                R2 = demux_r2, \n",
    "                nthreads  = 12,\n",
    "                output_folder = output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "wrapped-talent",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected bc_len: 16  Detected UMI len:  0\n",
      "updating progress every 3 minutes...\n",
      "number of read processed:0\n",
      "0\t0\t0\t0\t\n",
      "70000 reads processed, 70k reads/sec\n",
      "number of read processed: 70000\n",
      "time elapsed: 1050 milliseconds\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Demultiplexed BAM file sorted and indexed ...\n",
      "\n",
      "Using default value for barcode length (bc_length = 16) \n",
      "\n",
      "Generating mapping statistics per barcode\n",
      "\n",
      "Iterating through 5,000,000 reads at a time\n",
      "\n",
      "chunk1\n",
      "\n",
      "Completed generating mapping statistics per barcode, saved to /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/mapping_stats_per_barcode.csv\n",
      "\n",
      "\n",
      "The output tagged and sorted BAM file was sent to /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sorted_tagged_bam <- sc_atac_bam_tagging (inbam = aligned_bam, \n",
    "                       output_folder =  output_folder, \n",
    "                       bam_tags      = list(bc=\"CB\", mb=\"OX\"), \n",
    "                       nthreads      =  12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "hydraulic-subscriber",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "samtools was not located, so can't remove duplicates. Please make sure it is installed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "removed <- sc_atac_remove_duplicates(sorted_tagged_bam,\n",
    "                          output_folder = output_folder,)\n",
    "if (!isFALSE(removed))\n",
    "  sorted_tagged_bam <- removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "healthy-confirmation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder name is: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "Output BED file: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/fragments.bed\n"
     ]
    }
   ],
   "source": [
    "sc_atac_create_fragments(inbam = sorted_tagged_bam,\n",
    "                         output_folder = output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "angry-entry",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using genome size of1e+06\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sc_atac_peak_calling(inbam = sorted_tagged_bam,\n",
    "                     ref = reference,\n",
    "                     genome_size = NULL,\n",
    "                     output_folder = output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "economic-navigator",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`peak`, `tss` or `gene` feature_type is selected for feature input\n",
      "\n",
      "Creating Galignment object for the feature input ...\n",
      "\n",
      "hg38is a recognized organism. Using annotation files in repository.\n",
      "\n",
      "Generating TSS plot data\n",
      "\n",
      "Generating feature-barcode fragment count matrix\n",
      "\n",
      "Average no. of fragments per barcode: 1.34957764782326\n",
      "\n",
      "\n",
      "Raw feature matrix generated: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/unfiltered_feature_matrix.rds\n",
      "\n",
      "Calculating TSS enrichment scores\n",
      "\n",
      "Generating QC metrics for cells\n",
      "\n",
      "\n",
      "`none` function is used for cell calling ... \n",
      "\n",
      "nonewas not an implemented cell calling method\n",
      "\n",
      "cell called and stored in /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n",
      "Cells with less than 200 counts are filtered out.\n",
      "\n",
      "Number of cells to remove:1539\n",
      "\n",
      "No cells were filtered out since otherwise there would be too few left.\n",
      "\n",
      "features with less than 10 counts are filtered out.\n",
      "\n",
      "Number of features to remove:5\n",
      "\n",
      "No features were filtered out since otherwise there would be too few left.\n",
      "\n",
      "making sparse\n",
      "\n",
      "Sparse matrix generated\n",
      "\n",
      "Sparse count matrix is saved in\n",
      "/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/sparse_matrix.mtx\n",
      "\n",
      "Binary matrix is saved in:\n",
      "/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/binary_matrix.rds\n",
      "\n",
      "Computing feature QC metrics\n",
      "\n",
      "writing to csv\n",
      "\n",
      "SCE object is saved in:\n",
      "\t/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_SCEobject.rds\n",
      "\n",
      "\n",
      "\n",
      "processing file: rmd_report_skeleton.Rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/59                            \n",
      "2/59 [setup]                    \n",
      "3/59                            \n",
      "4/59 [unnamed-chunk-1]          \n",
      "5/59                            \n",
      "6/59 [inputs]                   \n",
      "7/59                            \n",
      "8/59 [unnamed-chunk-2]          \n",
      "9/59                            \n",
      "10/59 [unnamed-chunk-3]          \n",
      "11/59                            \n",
      "12/59 [unnamed-chunk-4]          \n",
      "13/59                            \n",
      "14/59 [alignment-stats-plot]     \n",
      "15/59                            \n",
      "16/59 [unnamed-chunk-5]          \n",
      "17/59                            \n",
      "18/59 [unnamed-chunk-6]          \n",
      "19/59                            \n",
      "20/59 [barcode-alignment-barplot]\n",
      "21/59                            \n",
      "22/59 [unnamed-chunk-7]          \n",
      "23/59                            \n",
      "24/59 [insert-size-dist]         \n",
      "25/59                            \n",
      "26/59 [unnamed-chunk-8]          \n",
      "27/59                            \n",
      "28/59 [frag-peak]                \n",
      "29/59                            \n",
      "30/59 [frag-mito]                \n",
      "31/59                            \n",
      "32/59 [frag-density]             \n",
      "33/59                            \n",
      "34/59 [barcode-alignment-dist]   \n",
      "35/59                            \n",
      "36/59 [percell-counts]           \n",
      "37/59                            \n",
      "38/59 [percell-features]         \n",
      "39/59                            \n",
      "40/59 [unnamed-chunk-9]          \n",
      "41/59                            \n",
      "42/59 [perfeature-counts]        \n",
      "43/59                            \n",
      "44/59 [perfeature-cells]         \n",
      "45/59                            \n",
      "46/59 [corr-cells]               \n",
      "47/59                            \n",
      "48/59 [corr-features]            \n",
      "49/59                            \n",
      "50/59 [overall-overlap]          \n",
      "51/59                            \n",
      "52/59 [anno-overlap-boxes]       \n",
      "53/59                            \n",
      "54/59 [tss-enrichment]           \n",
      "55/59                            \n",
      "56/59 [unnamed-chunk-10]         \n",
      "57/59                            \n",
      "58/59 [interactive-umap]         \n",
      "59/59                            \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: rmd_report_skeleton.knit.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/SGRNJ01/Public/Software/anaconda3/bin/pandoc +RTS -K512m -RTS rmd_report_skeleton.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/scPipe_atac_report.html --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --id-prefix section- --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmptznqHy/rmarkdown-str5755557d4cf.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Output created: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/scPipe_atac_report.html\n",
      "\n",
      "sc_atac_feature_counting completed in 23.8826081752777 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "features          <- file.path(output_folder, \"NA_peaks.narrowPeak\")\n",
    "\n",
    "sc_atac_feature_counting (fragment_file = file.path(output_folder, \"fragments.bed\"),\n",
    "                          feature_input = features, \n",
    "                          bam_tags      = list(bc=\"CB\", mb=\"OX\"), \n",
    "                          feature_type  = \"peak\",\n",
    "                          organism      = \"hg38\",\n",
    "                          cell_calling  = \"none\",\n",
    "                          min_uniq_frags = 0,\n",
    "                          min_frac_peak = 0,\n",
    "                          min_frac_promoter = 0,\n",
    "                          yieldsize     = 1000000,\n",
    "                          exclude_regions = TRUE,\n",
    "                          output_folder = output_folder,\n",
    "                          fix_chr       = \"none\",\n",
    "                          create_report = TRUE\n",
    "                          )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "popular-optimum",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caring-permit",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "occasional-pioneer",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "material-chuck",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "federal-jurisdiction",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving the output at location: \n",
      "/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No valid_barcode_file provided; no barcode error correction will occur.\n",
      "Total reads: 35000\n",
      "Total reads removed due to N's in barcodes: 0\n",
      "Total reads removed due to low quality barcodes: 0\n",
      "Total barcodes provided in FASTQ file: 16567\n",
      "time elapsed: 424 milliseconds\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ATAC-Seq mode is selected...\n",
      "\n",
      "Genome index location not specified. Looking for the index in/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n",
      "Genome index not found. Creating one in /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out ...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ==========     _____ _    _ ____  _____  ______          _____  \n",
      "        =====         / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "          =====      | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "            ====      \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "              ====    ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "        ==========   |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\n",
      "       Rsubread 2.14.2\n",
      "\n",
      "//================================= setting ==================================\\\\\n",
      "||                                                                            ||\n",
      "||                Index name : genome_index                                   ||\n",
      "||               Index space : base space                                     ||\n",
      "||               Index split : no-split                                       ||\n",
      "||          Repeat threshold : 100 repeats                                    ||\n",
      "||              Gapped index : no                                             ||\n",
      "||                                                                            ||\n",
      "||       Free / total memory : 59.4GB / 688.8GB                               ||\n",
      "||                                                                            ||\n",
      "||               Input files : 1 file in total                                ||\n",
      "||                             o small_chr21.fa                               ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//================================= Running ==================================\\\\\n",
      "||                                                                            ||\n",
      "|| Check the integrity of provided reference sequences ...                    ||\n",
      "|| No format issues were found                                                ||\n",
      "|| Scan uninformative subreads in reference sequences ...                     ||\n",
      "|| 3 uninformative subreads were found.                                       ||\n",
      "|| These subreads were excluded from index building.                          ||\n",
      "|| Estimate the index size...                                                 ||\n",
      "||    8%,   0 mins elapsed, rate=438.7k bps/s                                 ||\n",
      "||   16%,   0 mins elapsed, rate=804.2k bps/s                                 ||\n",
      "||   24%,   0 mins elapsed, rate=1152.3k bps/s                                ||\n",
      "||   33%,   0 mins elapsed, rate=1424.5k bps/s                                ||\n",
      "||   41%,   0 mins elapsed, rate=1710.8k bps/s                                ||\n",
      "||   49%,   0 mins elapsed, rate=1917.0k bps/s                                ||\n",
      "||   58%,   0 mins elapsed, rate=2099.7k bps/s                                ||\n",
      "||   66%,   0 mins elapsed, rate=2322.4k bps/s                                ||\n",
      "||   74%,   0 mins elapsed, rate=2464.5k bps/s                                ||\n",
      "||   83%,   0 mins elapsed, rate=2655.1k bps/s                                ||\n",
      "||   91%,   0 mins elapsed, rate=2885.5k bps/s                                ||\n",
      "|| 3.0 GB of memory is needed for index building.                             ||\n",
      "|| Build the index...                                                         ||\n",
      "||    8%,   0 mins elapsed, rate=38.2k bps/s                                  ||\n",
      "||   16%,   0 mins elapsed, rate=74.9k bps/s                                  ||\n",
      "||   24%,   0 mins elapsed, rate=111.1k bps/s                                 ||\n",
      "||   33%,   0 mins elapsed, rate=145.5k bps/s                                 ||\n",
      "||   41%,   0 mins elapsed, rate=180.0k bps/s                                 ||\n",
      "||   49%,   0 mins elapsed, rate=211.9k bps/s                                 ||\n",
      "||   58%,   0 mins elapsed, rate=242.5k bps/s                                 ||\n",
      "||   66%,   0 mins elapsed, rate=274.2k bps/s                                 ||\n",
      "||   74%,   0 mins elapsed, rate=302.6k bps/s                                 ||\n",
      "||   83%,   0 mins elapsed, rate=332.8k bps/s                                 ||\n",
      "||   91%,   0 mins elapsed, rate=364.6k bps/s                                 ||\n",
      "|| Save current index block...                                                ||\n",
      "||  [ 0.0% finished ]                                                         ||\n",
      "||  [ 10.0% finished ]                                                        ||\n",
      "||  [ 20.0% finished ]                                                        ||\n",
      "||  [ 30.0% finished ]                                                        ||\n",
      "||  [ 40.0% finished ]                                                        ||\n",
      "||  [ 50.0% finished ]                                                        ||\n",
      "||  [ 60.0% finished ]                                                        ||\n",
      "||  [ 70.0% finished ]                                                        ||\n",
      "||  [ 80.0% finished ]                                                        ||\n",
      "||  [ 90.0% finished ]                                                        ||\n",
      "||  [ 100.0% finished ]                                                       ||\n",
      "||                                                                            ||\n",
      "||                      Total running time: 0.2 minutes.                      ||\n",
      "||Index /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x ... ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No partial matches, checking for reads with non-matched barcodes.\n",
      "\n",
      "No reads found with non-matching barcodes.\n",
      "\n",
      "Outputted demultiplexing stats file to/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats//demultiplexing_stats.csv\n",
      "\n",
      "\n",
      "Output file name is not provided. Aligned reads are saved in /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/demux_completematch_small_chr21_R1_aligned.bam\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "        ==========     _____ _    _ ____  _____  ______          _____  \n",
      "        =====         / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "          =====      | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "            ====      \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "              ====    ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "        ==========   |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\n",
      "       Rsubread 2.14.2\n",
      "\n",
      "//================================= setting ==================================\\\\\n",
      "||                                                                            ||\n",
      "|| Function      : Read alignment (DNA-Seq)                                   ||\n",
      "|| Input file 1  : demux_completematch_small_chr21_R1.fastq.gz                ||\n",
      "|| Input file 2  : demux_completematch_small_chr21_R3.fastq.gz                ||\n",
      "|| Output file   : demux_completematch_small_chr21_R1_aligned.bam (BAM),  ... ||\n",
      "|| Index name    : genome_index                                               ||\n",
      "||                                                                            ||\n",
      "||                    ------------------------------------                    ||\n",
      "||                                                                            ||\n",
      "||                               Threads : 8                                  ||\n",
      "||                          Phred offset : 33                                 ||\n",
      "||               # of extracted subreads : 10                                 ||\n",
      "||                        Min read1 vote : 3                                  ||\n",
      "||                        Min read2 vote : 1                                  ||\n",
      "||                     Max fragment size : 600                                ||\n",
      "||                     Min fragment size : 50                                 ||\n",
      "||                        Max mismatches : 3                                  ||\n",
      "||                      Max indel length : 5                                  ||\n",
      "||            Report multi-mapping reads : yes                                ||\n",
      "|| Max alignments per multi-mapping read : 1                                  ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//================ Running (06-Sep-2023 13:46:33, pid=22357) =================\\\\\n",
      "||                                                                            ||\n",
      "|| Check the input reads.                                                     ||\n",
      "|| The input file contains base space reads.                                  ||\n",
      "|| Initialise the memory objects.                                             ||\n",
      "|| Estimate the mean read length.                                             ||\n",
      "|| The range of Phred scores observed in the data is [2,37]                   ||\n",
      "|| Create the output BAM file.                                                ||\n",
      "|| Check the index.                                                           ||\n",
      "|| Init the voting space.                                                     ||\n",
      "|| Global environment is initialised.                                         ||\n",
      "|| Load the 1-th index block...                                               ||\n",
      "|| The index block has been loaded.                                           ||\n",
      "|| Start read mapping in chunk.                                               ||\n",
      "||   Estimated fragment length : 123 bp                                       ||\n",
      "||   98% completed, 0.5 mins elapsed, rate=0.9k fragments per second          ||\n",
      "||                                                                            ||\n",
      "||                           Completed successfully.                          ||\n",
      "||                                                                            ||\n",
      "\\\\====================================    ====================================//\n",
      "\n",
      "//================================   Summary =================================\\\\\n",
      "||                                                                            ||\n",
      "||             Total fragments : 35,000                                       ||\n",
      "||                      Mapped : 18,851 (53.9%)                               ||\n",
      "||             Uniquely mapped : 18,057                                       ||\n",
      "||               Multi-mapping : 794                                          ||\n",
      "||                                                                            ||\n",
      "||                    Unmapped : 16,149                                       ||\n",
      "||                                                                            ||\n",
      "||             Properly paired : 14,524                                       ||\n",
      "||         Not properly paired : 4,327                                        ||\n",
      "||                   Singleton : 1,458                                        ||\n",
      "||                    Chimeric : 0                                            ||\n",
      "||       Unexpected strandness : 114                                          ||\n",
      "||  Unexpected fragment length : 180                                          ||\n",
      "||       Unexpected read order : 2,575                                        ||\n",
      "||                                                                            ||\n",
      "||                      Indels : 197                                          ||\n",
      "||                                                                            ||\n",
      "||                Running time : 0.6 minutes                                  ||\n",
      "||                                                                            ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "Detected bc_len: 16  Detected UMI len:  0\n",
      "updating progress every 3 minutes...\n",
      "number of read processed:0\n",
      "0\t0\t0\t0\t\n",
      "70000 reads processed, 70k reads/sec\n",
      "number of read processed: 70000\n",
      "time elapsed: 1071 milliseconds\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Demultiplexed BAM file sorted and indexed ...\n",
      "\n",
      "Using default value for barcode length (bc_length = 16) \n",
      "\n",
      "Generating mapping statistics per barcode\n",
      "\n",
      "Iterating through 5,000,000 reads at a time\n",
      "\n",
      "chunk1\n",
      "\n",
      "Completed generating mapping statistics per barcode, saved to /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/mapping_stats_per_barcode.csv\n",
      "\n",
      "\n",
      "The output tagged and sorted BAM file was sent to /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n",
      "samtools was not located, so can't remove duplicates. Please make sure it is installed.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output folder name is: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "Output BED file: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/fragments.bed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using genome size of1e+06\n",
      "\n",
      "`peak`, `tss` or `gene` feature_type is selected for feature input\n",
      "\n",
      "Creating Galignment object for the feature input ...\n",
      "\n",
      "hg38is a recognized organism. Using annotation files in repository.\n",
      "\n",
      "Generating TSS plot data\n",
      "\n",
      "Generating feature-barcode fragment count matrix\n",
      "\n",
      "Average no. of fragments per barcode: 1.34957764782326\n",
      "\n",
      "\n",
      "Raw feature matrix generated: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/unfiltered_feature_matrix.rds\n",
      "\n",
      "Calculating TSS enrichment scores\n",
      "\n",
      "Generating QC metrics for cells\n",
      "\n",
      "\n",
      "`none` function is used for cell calling ... \n",
      "\n",
      "nonewas not an implemented cell calling method\n",
      "\n",
      "cell called and stored in /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\n",
      "\n",
      "Cells with less than 200 counts are filtered out.\n",
      "\n",
      "Number of cells to remove:1539\n",
      "\n",
      "No cells were filtered out since otherwise there would be too few left.\n",
      "\n",
      "features with less than 10 counts are filtered out.\n",
      "\n",
      "Number of features to remove:5\n",
      "\n",
      "No features were filtered out since otherwise there would be too few left.\n",
      "\n",
      "making sparse\n",
      "\n",
      "Sparse matrix generated\n",
      "\n",
      "Sparse count matrix is saved in\n",
      "/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/sparse_matrix.mtx\n",
      "\n",
      "Binary matrix is saved in:\n",
      "/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/binary_matrix.rds\n",
      "\n",
      "Computing feature QC metrics\n",
      "\n",
      "writing to csv\n",
      "\n",
      "SCE object is saved in:\n",
      "\t/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_SCEobject.rds\n",
      "\n",
      "\n",
      "\n",
      "processing file: rmd_report_skeleton.Rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/59                            \n",
      "2/59 [setup]                    \n",
      "3/59                            \n",
      "4/59 [unnamed-chunk-1]          \n",
      "5/59                            \n",
      "6/59 [inputs]                   \n",
      "7/59                            \n",
      "8/59 [unnamed-chunk-2]          \n",
      "9/59                            \n",
      "10/59 [unnamed-chunk-3]          \n",
      "11/59                            \n",
      "12/59 [unnamed-chunk-4]          \n",
      "13/59                            \n",
      "14/59 [alignment-stats-plot]     \n",
      "15/59                            \n",
      "16/59 [unnamed-chunk-5]          \n",
      "17/59                            \n",
      "18/59 [unnamed-chunk-6]          \n",
      "19/59                            \n",
      "20/59 [barcode-alignment-barplot]\n",
      "21/59                            \n",
      "22/59 [unnamed-chunk-7]          \n",
      "23/59                            \n",
      "24/59 [insert-size-dist]         \n",
      "25/59                            \n",
      "26/59 [unnamed-chunk-8]          \n",
      "27/59                            \n",
      "28/59 [frag-peak]                \n",
      "29/59                            \n",
      "30/59 [frag-mito]                \n",
      "31/59                            \n",
      "32/59 [frag-density]             \n",
      "33/59                            \n",
      "34/59 [barcode-alignment-dist]   \n",
      "35/59                            \n",
      "36/59 [percell-counts]           \n",
      "37/59                            \n",
      "38/59 [percell-features]         \n",
      "39/59                            \n",
      "40/59 [unnamed-chunk-9]          \n",
      "41/59                            \n",
      "42/59 [perfeature-counts]        \n",
      "43/59                            \n",
      "44/59 [perfeature-cells]         \n",
      "45/59                            \n",
      "46/59 [corr-cells]               \n",
      "47/59                            \n",
      "48/59 [corr-features]            \n",
      "49/59                            \n",
      "50/59 [overall-overlap]          \n",
      "51/59                            \n",
      "52/59 [anno-overlap-boxes]       \n",
      "53/59                            \n",
      "54/59 [tss-enrichment]           \n",
      "55/59                            \n",
      "56/59 [unnamed-chunk-10]         \n",
      "57/59                            \n",
      "58/59 [interactive-umap]         \n",
      "59/59                            \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: rmd_report_skeleton.knit.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/SGRNJ01/Public/Software/anaconda3/bin/pandoc +RTS -K512m -RTS rmd_report_skeleton.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/scPipe_atac_report.html --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --id-prefix section- --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmptznqHy/rmarkdown-str57555b105ced.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Output created: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/scPipe_atac_report.html\n",
      "\n",
      "sc_atac_feature_counting completed in 27.1890134811401 seconds\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  ..@ i       : int [1:1582] 4 7 4 4 4 4 4 4 4 1 ...\n",
      "  ..@ p       : int [1:1540] 0 1 2 3 4 5 6 7 8 9 ...\n",
      "  ..@ Dim     : int [1:2] 12 1539\n",
      "  ..@ Dimnames:List of 2\n",
      "  .. ..$ feature: chr [1:12] \"chr21:284484-285174\" \"chr21:287515-288092\" \"chr21:288148-288891\" \"chr21:289012-289228\" ...\n",
      "  .. ..$ barcode: chr [1:1539] \"AAACGAACAGTAAGCG\" \"AAACGAACATGGATGG\" \"AAACGAATCGGACGAA\" \"AAACGAATCTGTGTGA\" ...\n",
      "  ..@ x       : num [1:1582] 1 1 1 1 1 1 2 1 1 1 ...\n",
      "  ..@ factors : list()\n",
      "Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  ..@ i       : int [1:1582] 4 7 4 4 4 4 4 4 4 1 ...\n",
      "  ..@ p       : int [1:1540] 0 1 2 3 4 5 6 7 8 9 ...\n",
      "  ..@ Dim     : int [1:2] 12 1539\n",
      "  ..@ Dimnames:List of 2\n",
      "  .. ..$ feature: chr [1:12] \"chr21:284484-285174\" \"chr21:287515-288092\" \"chr21:288148-288891\" \"chr21:289012-289228\" ...\n",
      "  .. ..$ barcode: chr [1:1539] \"AAACGAACAGTAAGCG\" \"AAACGAACATGGATGG\" \"AAACGAATCGGACGAA\" \"AAACGAATCTGTGTGA\" ...\n",
      "  ..@ x       : num [1:1582] 1 1 1 1 1 1 2 1 1 1 ...\n",
      "  ..@ factors : list()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SCE object is saved in:\n",
      "\t/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_SCEobject.rds\n",
      "\n",
      "\n",
      "\n",
      "processing file: rmd_report_skeleton.Rmd\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/59                            \n",
      "2/59 [setup]                    \n",
      "3/59                            \n",
      "4/59 [unnamed-chunk-1]          \n",
      "5/59                            \n",
      "6/59 [inputs]                   \n",
      "7/59                            \n",
      "8/59 [unnamed-chunk-2]          \n",
      "9/59                            \n",
      "10/59 [unnamed-chunk-3]          \n",
      "11/59                            \n",
      "12/59 [unnamed-chunk-4]          \n",
      "13/59                            \n",
      "14/59 [alignment-stats-plot]     \n",
      "15/59                            \n",
      "16/59 [unnamed-chunk-5]          \n",
      "17/59                            \n",
      "18/59 [unnamed-chunk-6]          \n",
      "19/59                            \n",
      "20/59 [barcode-alignment-barplot]\n",
      "21/59                            \n",
      "22/59 [unnamed-chunk-7]          \n",
      "23/59                            \n",
      "24/59 [insert-size-dist]         \n",
      "25/59                            \n",
      "26/59 [unnamed-chunk-8]          \n",
      "27/59                            \n",
      "28/59 [frag-peak]                \n",
      "29/59                            \n",
      "30/59 [frag-mito]                \n",
      "31/59                            \n",
      "32/59 [frag-density]             \n",
      "33/59                            \n",
      "34/59 [barcode-alignment-dist]   \n",
      "35/59                            \n",
      "36/59 [percell-counts]           \n",
      "37/59                            \n",
      "38/59 [percell-features]         \n",
      "39/59                            \n",
      "40/59 [unnamed-chunk-9]          \n",
      "41/59                            \n",
      "42/59 [perfeature-counts]        \n",
      "43/59                            \n",
      "44/59 [perfeature-cells]         \n",
      "45/59                            \n",
      "46/59 [corr-cells]               \n",
      "47/59                            \n",
      "48/59 [corr-features]            \n",
      "49/59                            \n",
      "50/59 [overall-overlap]          \n",
      "51/59                            \n",
      "52/59 [anno-overlap-boxes]       \n",
      "53/59                            \n",
      "54/59 [tss-enrichment]           \n",
      "55/59                            \n",
      "56/59 [unnamed-chunk-10]         \n",
      "57/59                            \n",
      "58/59 [interactive-umap]         \n",
      "59/59                            \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "output file: rmd_report_skeleton.knit.md\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/SGRNJ01/Public/Software/anaconda3/bin/pandoc +RTS -K512m -RTS rmd_report_skeleton.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/scPipe_atac_report.html --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmarkdown/lua/latex-div.lua --self-contained --variable bs3=TRUE --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --id-prefix section- --variable theme=bootstrap --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/RtmptznqHy/rmarkdown-str575548bf4cda.html \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Output created: /SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out/scPipe_atac_stats/scPipe_atac_report.html\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sample <- \"small_chr21\"\n",
    "r1 <- \"/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata/small_chr21_R1.fastq.gz\"\n",
    "r2 <- \"/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata/small_chr21_R2.fastq.gz\"\n",
    "r3 <- \"/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata/small_chr21_R3.fastq.gz\"\n",
    "thread <- 8\n",
    "output_folder <- \"/SGRNJ06/randd/USER/cjj/celedev/atac/celescope_atac/20230905_10x/test/test_out\"\n",
    "reference <- \"/SGRNJ01/Public/Software/conda_env/scpipe/lib/R/library/scPipe/extdata/small_chr21.fa\"\n",
    "\n",
    "# convert format\n",
    "sc_atac_trim_barcode (r1            = r1, \n",
    "                      r2            = r3, \n",
    "                      bc_file       = r2, \n",
    "                      rmN           = FALSE,\n",
    "                      rmlow         = FALSE,\n",
    "                      output_folder = output_folder)\n",
    "\n",
    "# Aligning reads to a reference genome\n",
    "demux_r1 <- file.path(output_folder, paste0(\"demux_completematch_\", sample, \"_R1.fastq.gz\"))\n",
    "demux_r2 <- file.path(output_folder, paste0(\"demux_completematch_\", sample, \"_R3.fastq.gz\"))\n",
    "aligned_bam <- sc_aligning(ref = reference,\n",
    "                R1 = demux_r1,\n",
    "                R2 = demux_r2,\n",
    "                nthreads  = thread,\n",
    "                output_folder = output_folder)\n",
    "\n",
    "# Demultiplexing the BAM file\n",
    "sorted_tagged_bam <- sc_atac_bam_tagging (inbam = aligned_bam,\n",
    "                       output_folder =  output_folder,\n",
    "                       bam_tags      = list(bc=\"CB\", mb=\"OX\"),\n",
    "                       nthreads      =  thread)\n",
    "\n",
    "# Remove duplicates\n",
    "removed <- sc_atac_remove_duplicates(sorted_tagged_bam,\n",
    "                          output_folder = output_folder,)\n",
    "if (!isFALSE(removed))\n",
    "  sorted_tagged_bam <- removed\n",
    "\n",
    "# Gemerating a fragment file\n",
    "sc_atac_create_fragments(inbam = sorted_tagged_bam,\n",
    "                         output_folder = output_folder)\n",
    "\n",
    "# Peak calling\n",
    "sc_atac_peak_calling(inbam = sorted_tagged_bam,\n",
    "                     ref = reference,\n",
    "                     genome_size = NULL,\n",
    "                     output_folder = output_folder)\n",
    "\n",
    "# Assigning reads to features and feature counting\n",
    "features          <- file.path(output_folder, \"NA_peaks.narrowPeak\")\n",
    "sc_atac_feature_counting (fragment_file = file.path(output_folder, \"fragments.bed\"),\n",
    "                          feature_input = features,\n",
    "                          bam_tags      = list(bc=\"CB\", mb=\"OX\"),\n",
    "                          feature_type  = \"peak\",\n",
    "                          organism      = \"hg38\",\n",
    "                          cell_calling  = \"none\",\n",
    "                          min_uniq_frags = 0,\n",
    "                          min_frac_peak = 0,\n",
    "                          min_frac_promoter = 0,\n",
    "                          yieldsize     = 1000000,\n",
    "                          exclude_regions = TRUE,\n",
    "                          output_folder = output_folder,\n",
    "                          fix_chr       = \"none\",\n",
    "                          create_report = TRUE\n",
    "                          )\n",
    "\n",
    "feature_matrix <- readRDS(file.path(output_folder, \"unfiltered_feature_matrix.rds\"))\n",
    "dplyr::glimpse(feature_matrix)\n",
    "sparseM <- readRDS(file.path(output_folder, \"sparse_matrix.rds\"))\n",
    "dplyr::glimpse(sparseM)\n",
    "\n",
    "# Generating the Single-cell Experiment (SCE) object\n",
    "sce <- sc_atac_create_sce(input_folder = output_folder,\n",
    "                   organism     = \"hg38\",\n",
    "                   feature_type = \"peak\",\n",
    "                   pheno_data   = NULL,\n",
    "                   report       = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "directed-index",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scpipe",
   "language": "R",
   "name": "scpipe"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
