{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "educational-blanket",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import copy\n",
    "import os\n",
    "import subprocess\n",
    "from Bio.Seq import Seq\n",
    "from celescope.tools import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "naval-startup",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Auto():\n",
    "    \"\"\"\n",
    "    threshold = top {percentile}% cell count / coef\n",
    "    count is usually UMI count.\n",
    "    >>> array = [50] * 100 + [30] * 100 + [10] * 100 + [4] * 100\n",
    "    >>> Auto(array, coef=10).run()\n",
    "    5\n",
    "    >>> Auto(array, percentile=70, coef=3).run()\n",
    "    10\n",
    "    >>> Auto(array, percentile=50, coef=10, expected_cell_num=100).run()\n",
    "    5\n",
    "    >>> Auto([1, 2, 20, 30, 40], expected_cell_num=4, percentile=50, coef=10).run()\n",
    "    2\n",
    "    \"\"\"\n",
    "    def __init__(self, array, percentile=99, coef=3, expected_cell_num=None, **kwargs):\n",
    "        self.array = [x for x in array if x > 0 ]\n",
    "        self.percentile = percentile\n",
    "        self.coef = int(coef)\n",
    "        self.expected_cell_num = expected_cell_num\n",
    "        self.kwargs = kwargs\n",
    "    \n",
    "    def run(self):\n",
    "        array = self.array\n",
    "        if not array:\n",
    "            return 1\n",
    "\n",
    "        if not self.expected_cell_num:\n",
    "            expected_cell_num = len(array)\n",
    "        else:\n",
    "            expected_cell_num = self.expected_cell_num\n",
    "            if expected_cell_num > len(array):\n",
    "                print('Warning: expected_cell_num > len(array)')\n",
    "                expected_cell_num = len(array)\n",
    "                      \n",
    "        sorted_counts = sorted(array, reverse=True)\n",
    "        count_cell_percentile = np.percentile(sorted_counts[:expected_cell_num], self.percentile)\n",
    "        threshold = int(count_cell_percentile / self.coef)\n",
    "\n",
    "        return threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "cleared-round",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reversed_compl(seq):\n",
    "    return str(Seq(seq).reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "fresh-alignment",
   "metadata": {},
   "outputs": [],
   "source": [
    "def target_cell_calling(df_UMI_sum, expected_target_cell_num=3000, target_barcodes=None, weight=6, coef=5, \n",
    "    percentile=80, umi_col='umis'):\n",
    "    \"\"\"\n",
    "    Args:\n",
    "        df_UMI_sum: A dataframe with columns highest umi's contig and UMI.\n",
    "    \n",
    "    Returns:\n",
    "        umi_threhold: int\n",
    "        target_cell_barcodes: list\n",
    "    >>> df_UMI_sum = pd.DataFrame({\"contig_id\": [\"A\", \"B\", \"C\", \"D\", \"E\"], \"UMI\": [1, 2, 1, 30, 40]})\n",
    "    >>> umi_threshold, target_cell_barcodes = target_cell_calling(df_UMI_sum, expected_target_cell_num=5, percentile=80, coef=5, target_barcodes=[\"A\", \"C\"])\n",
    "    >>> umi_threshold == 3\n",
    "    True\n",
    "    >>> target_cell_barcodes == {'A', 'C', 'D', 'E'}\n",
    "    True\n",
    "    \"\"\"\n",
    "    if target_barcodes != None:\n",
    "        target_barcodes = {reversed_compl(i) for i in target_barcodes}\n",
    "    umi_threshold = Auto(list(df_UMI_sum[umi_col]), expected_cell_num=expected_target_cell_num, coef=coef, percentile=percentile).run()\n",
    "\n",
    "    # avoid change the original dataframe\n",
    "    df_temp = df_UMI_sum.copy()\n",
    "    if target_barcodes:\n",
    "        df_temp[umi_col] = df_temp.apply(\n",
    "            lambda row:  row[umi_col] * weight if row['barcode'] in target_barcodes else row[umi_col], axis=1)\n",
    "             \n",
    "    target_contigs = set(df_temp.loc[df_temp[umi_col] >= umi_threshold].contig_id)\n",
    "\n",
    "    return target_contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "tribal-filling",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Personal/chenjunjie/.local/lib/python3.7/site-packages/Bio/Seq.py:2983: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "        df = pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/kemai/20220517merTRflilter/sc-1/03.assemble/assemble/sc-1_contig.csv', sep='\\t', header=None)\n",
    "        df.columns = ['barcode', 'is_cell', 'contig_id', 'high_confidence', 'length', 'chain', 'v_gene', 'd_gene', 'j_gene', 'c_gene', 'full_length', 'productive', 'cdr3', 'cdr3_nt', 'reads', 'umis']\n",
    "        df['d_gene'] = df['d_gene'].apply(lambda x: x.split('(')[0] if not x == '*' else 'None')\n",
    "        df['c_gene'] = df['c_gene'].apply(lambda x: x.split('(')[0] if not x == '*' else 'None')\n",
    "        df['cdr3'] = df['cdr3_nt'].apply(lambda x: 'None' if \"*\" in str(Seq(x).translate()) or not len(x)%3==0 else str(Seq(x).translate()))\n",
    "        df['productive'] = df['cdr3'].apply(lambda x: False if x=='None' else True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "id": "white-cardiff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.sort_values(by='umis', ascending=False, inplace=True)\n",
    "df_chain_heavy = df[df['chain']=='IGH']\n",
    "df_chain_light = df[(df['chain']=='IGK') | (df['chain']=='IGL')]\n",
    "df_chain_heavy = df_chain_heavy.drop_duplicates(['barcode'])\n",
    "df_chain_light = df_chain_light.drop_duplicates(['barcode'])\n",
    "df_for_clono = pd.concat([df_chain_heavy, df_chain_light], ignore_index=True)\n",
    "trust_report = pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/kemai/20220517merTRflilter/sc-1/03.assemble/assemble/trust_filter_report.out', sep='\\t')\n",
    "correct_cdr3 = set(df_for_clono.cdr3).intersection(set(trust_report.CDR3aa))\n",
    "correct_cdr3 = [i for i in correct_cdr3 if i.startswith('C')]\n",
    "correct_cdr3 = [i for i in correct_cdr3 if len(i)>=5]\n",
    "correct_cdr3 = [i for i in correct_cdr3 if 'UAG' or 'UAA' or 'UGA' not in i]\n",
    "df_for_clono = df_for_clono[df_for_clono['cdr3'].isin(correct_cdr3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "surprised-interpretation",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chain_heavy = df_for_clono[df_for_clono['chain']=='IGH']\n",
    "df_chain_light = df_for_clono[(df_for_clono['chain']=='IGK') | (df_for_clono['chain']=='IGL')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "convertible-discrimination",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_barcodes = utils.read_one_col('/SGRNJ06/randd/USER/cjj/TESTDATA/test_trust/barcodelist.csv')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "boolean-lexington",
   "metadata": {},
   "outputs": [],
   "source": [
    "#target_barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "viral-dakota",
   "metadata": {},
   "outputs": [],
   "source": [
    "        filtered_congtigs_id = set()\n",
    "        for _df in [df_chain_heavy, df_chain_light]:\n",
    "            target_contigs = target_cell_calling(\n",
    "            _df, \n",
    "            #expected_target_cell_num=3000, \n",
    "            #target_barcodes=None,\n",
    "            expected_target_cell_num = len(target_barcodes),\n",
    "            target_barcodes = target_barcodes,\n",
    "            weight=6, coef=5, \n",
    "            percentile=85,\n",
    "            )\n",
    "            filtered_congtigs_id = filtered_congtigs_id | target_contigs       \n",
    "        \n",
    "        df_for_clono = df_for_clono[df_for_clono.contig_id.isin(filtered_congtigs_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "periodic-paragraph",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGATACGTTCAGGAGGTTGTTCCA</td>\n",
       "      <td>True</td>\n",
       "      <td>TGATACGTTCAGGAGGTTGTTCCA_899</td>\n",
       "      <td>True</td>\n",
       "      <td>517</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-20*01</td>\n",
       "      <td>IGHD1-20*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHG2B*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARGITGIDYW</td>\n",
       "      <td>TGTGCGAGGGGTATAACTGGAATTGACTACTGG</td>\n",
       "      <td>242751</td>\n",
       "      <td>105937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TAGACGGAGTCGCTATTCGAGCGT</td>\n",
       "      <td>True</td>\n",
       "      <td>TAGACGGAGTCGCTATTCGAGCGT_1286</td>\n",
       "      <td>True</td>\n",
       "      <td>605</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-23*04</td>\n",
       "      <td>IGHD7-27*01</td>\n",
       "      <td>IGHJ2*01</td>\n",
       "      <td>IGHG1*01</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAKDRNWGYEYFDLW</td>\n",
       "      <td>TGTGCGAAAGATCGAAACTGGGGATACGAGTACTTCGATCTCTGG</td>\n",
       "      <td>146639</td>\n",
       "      <td>79062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTTGTCGGTCTCGGTTTGCATAGT</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTGTCGGTCTCGGTTTGCATAGT_525</td>\n",
       "      <td>True</td>\n",
       "      <td>804</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-20*01</td>\n",
       "      <td>IGHD3-9*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHG2B*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARDLTLFDYW</td>\n",
       "      <td>TGTGCGAGAGATCTAACTCTTTTTGACTACTGG</td>\n",
       "      <td>138706</td>\n",
       "      <td>71297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTCTGTGTGGTCGTGTCGATGTTT</td>\n",
       "      <td>True</td>\n",
       "      <td>TTCTGTGTGGTCGTGTCGATGTTT_5973</td>\n",
       "      <td>True</td>\n",
       "      <td>522</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-39*01</td>\n",
       "      <td>IGHD1-7*01</td>\n",
       "      <td>IGHJ2*01</td>\n",
       "      <td>IGHA*09</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CASLYNWNYDWFDPWGLW</td>\n",
       "      <td>TGTGCGAGCCTGTATAACTGGAACTACGACTGGTTCGACCCCTGGG...</td>\n",
       "      <td>74544</td>\n",
       "      <td>45872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTGTGTCGGCTAACTCTCGAAGTG</td>\n",
       "      <td>True</td>\n",
       "      <td>GTGTGTCGGCTAACTCTCGAAGTG_2191</td>\n",
       "      <td>True</td>\n",
       "      <td>536</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-48*03</td>\n",
       "      <td>IGHD3-22*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHA*09</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAREADWLVHYYFDYW</td>\n",
       "      <td>TGTGCGAGAGAGGCGGACTGGCTGGTTCATTACTACTTTGACTACTGG</td>\n",
       "      <td>55476</td>\n",
       "      <td>39029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10428</th>\n",
       "      <td>TGACAGACGCACTGTCTGTCTATC</td>\n",
       "      <td>True</td>\n",
       "      <td>TGACAGACGCACTGTCTGTCTATC_2700</td>\n",
       "      <td>True</td>\n",
       "      <td>516</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNWPPTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC</td>\n",
       "      <td>207</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10429</th>\n",
       "      <td>TCCGTCTTTCAGGAGGTGCGATCT</td>\n",
       "      <td>True</td>\n",
       "      <td>TCCGTCTTTCAGGAGGTGCGATCT_3695</td>\n",
       "      <td>True</td>\n",
       "      <td>503</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNWPPTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC</td>\n",
       "      <td>189</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10430</th>\n",
       "      <td>TCGAAGTGTCTCACGGTATGTGGC</td>\n",
       "      <td>True</td>\n",
       "      <td>TCGAAGTGTCTCACGGTATGTGGC_2885</td>\n",
       "      <td>True</td>\n",
       "      <td>522</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNWPPTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC</td>\n",
       "      <td>229</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10431</th>\n",
       "      <td>TCCGTCTTGAGGATGGGTGTGTCG</td>\n",
       "      <td>True</td>\n",
       "      <td>TCCGTCTTGAGGATGGGTGTGTCG_2513</td>\n",
       "      <td>True</td>\n",
       "      <td>523</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNWPPTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC</td>\n",
       "      <td>172</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10433</th>\n",
       "      <td>GAATCTGTTCAGATTCTTCGCACC</td>\n",
       "      <td>True</td>\n",
       "      <td>GAATCTGTTCAGATTCTTCGCACC_2580</td>\n",
       "      <td>True</td>\n",
       "      <td>525</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC*03</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNWPPTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC</td>\n",
       "      <td>190</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8660 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        barcode  is_cell                      contig_id  \\\n",
       "0      TGATACGTTCAGGAGGTTGTTCCA     True   TGATACGTTCAGGAGGTTGTTCCA_899   \n",
       "1      TAGACGGAGTCGCTATTCGAGCGT     True  TAGACGGAGTCGCTATTCGAGCGT_1286   \n",
       "2      GTTGTCGGTCTCGGTTTGCATAGT     True   GTTGTCGGTCTCGGTTTGCATAGT_525   \n",
       "3      TTCTGTGTGGTCGTGTCGATGTTT     True  TTCTGTGTGGTCGTGTCGATGTTT_5973   \n",
       "4      GTGTGTCGGCTAACTCTCGAAGTG     True  GTGTGTCGGCTAACTCTCGAAGTG_2191   \n",
       "...                         ...      ...                            ...   \n",
       "10428  TGACAGACGCACTGTCTGTCTATC     True  TGACAGACGCACTGTCTGTCTATC_2700   \n",
       "10429  TCCGTCTTTCAGGAGGTGCGATCT     True  TCCGTCTTTCAGGAGGTGCGATCT_3695   \n",
       "10430  TCGAAGTGTCTCACGGTATGTGGC     True  TCGAAGTGTCTCACGGTATGTGGC_2885   \n",
       "10431  TCCGTCTTGAGGATGGGTGTGTCG     True  TCCGTCTTGAGGATGGGTGTGTCG_2513   \n",
       "10433  GAATCTGTTCAGATTCTTCGCACC     True  GAATCTGTTCAGATTCTTCGCACC_2580   \n",
       "\n",
       "       high_confidence  length chain       v_gene       d_gene    j_gene  \\\n",
       "0                 True     517   IGH  IGHV3-20*01  IGHD1-20*01  IGHJ4*02   \n",
       "1                 True     605   IGH  IGHV3-23*04  IGHD7-27*01  IGHJ2*01   \n",
       "2                 True     804   IGH  IGHV3-20*01   IGHD3-9*01  IGHJ4*02   \n",
       "3                 True     522   IGH  IGHV4-39*01   IGHD1-7*01  IGHJ2*01   \n",
       "4                 True     536   IGH  IGHV3-48*03  IGHD3-22*01  IGHJ4*02   \n",
       "...                ...     ...   ...          ...          ...       ...   \n",
       "10428             True     516   IGK  IGKV3-11*01         None  IGKJ1*01   \n",
       "10429             True     503   IGK  IGKV3-11*01         None  IGKJ1*01   \n",
       "10430             True     522   IGK  IGKV3-11*01         None  IGKJ1*01   \n",
       "10431             True     523   IGK  IGKV3-11*01         None  IGKJ1*01   \n",
       "10433             True     525   IGK  IGKV3-11*01         None  IGKJ1*01   \n",
       "\n",
       "          c_gene  full_length  productive                cdr3  \\\n",
       "0      IGHG2B*03         True        True         CARGITGIDYW   \n",
       "1       IGHG1*01         True        True     CAKDRNWGYEYFDLW   \n",
       "2      IGHG2B*03         True        True         CARDLTLFDYW   \n",
       "3        IGHA*09         True        True  CASLYNWNYDWFDPWGLW   \n",
       "4        IGHA*09         True        True    CAREADWLVHYYFDYW   \n",
       "...          ...          ...         ...                 ...   \n",
       "10428    IGKC*03         True        True         CQQRSNWPPTF   \n",
       "10429    IGKC*03         True        True         CQQRSNWPPTF   \n",
       "10430    IGKC*03         True        True         CQQRSNWPPTF   \n",
       "10431    IGKC*03         True        True         CQQRSNWPPTF   \n",
       "10433    IGKC*03         True        True         CQQRSNWPPTF   \n",
       "\n",
       "                                                 cdr3_nt   reads    umis  \n",
       "0                      TGTGCGAGGGGTATAACTGGAATTGACTACTGG  242751  105937  \n",
       "1          TGTGCGAAAGATCGAAACTGGGGATACGAGTACTTCGATCTCTGG  146639   79062  \n",
       "2                      TGTGCGAGAGATCTAACTCTTTTTGACTACTGG  138706   71297  \n",
       "3      TGTGCGAGCCTGTATAACTGGAACTACGACTGGTTCGACCCCTGGG...   74544   45872  \n",
       "4       TGTGCGAGAGAGGCGGACTGGCTGGTTCATTACTACTTTGACTACTGG   55476   39029  \n",
       "...                                                  ...     ...     ...  \n",
       "10428                  TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC     207     153  \n",
       "10429                  TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC     189     153  \n",
       "10430                  TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC     229     153  \n",
       "10431                  TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC     172     153  \n",
       "10433                  TGTCAGCAGCGTAGCAACTGGCCTCCGACGTTC     190     153  \n",
       "\n",
       "[8660 rows x 16 columns]"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_clono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "smaller-virginia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'Cells With Productive V-J Spanning Pair',\n",
       "  'value': 4048,\n",
       "  'total': 4612},\n",
       " {'name': 'Cells With Productive V-J Spanning (IGH, IGL) Pair',\n",
       "  'value': 0,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With Productive V-J Spanning (IGH, IGK) Pair',\n",
       "  'value': 4048,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGH Contig',\n",
       "  'value': 4413,\n",
       "  'total': 4612,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGH contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGH Contig',\n",
       "  'value': 4413,\n",
       "  'total': 4612},\n",
       " {'name': 'Cells With V-J Spanning IGH Contig',\n",
       "  'value': 4413,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGH\"},\n",
       " {'name': 'Cells With Productive IGH Contig',\n",
       "  'value': 4413,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGL Contig',\n",
       "  'value': 0,\n",
       "  'total': 4612,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGL contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGL Contig', 'value': 0, 'total': 4612},\n",
       " {'name': 'Cells With V-J Spanning IGL Contig',\n",
       "  'value': 0,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGL\"},\n",
       " {'name': 'Cells With Productive IGL Contig',\n",
       "  'value': 0,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGK Contig',\n",
       "  'value': 4247,\n",
       "  'total': 4612,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGK contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGK Contig',\n",
       "  'value': 4247,\n",
       "  'total': 4612},\n",
       " {'name': 'Cells With V-J Spanning IGK Contig',\n",
       "  'value': 4247,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGK\"},\n",
       " {'name': 'Cells With Productive IGK Contig',\n",
       "  'value': 4247,\n",
       "  'total': 4612,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"}]"
      ]
     },
     "execution_count": 351,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_vj_annot(df_for_clono, ['IGH', 'IGL', 'IGK'], ['IGH_IGL', 'IGH_IGK'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "administrative-western",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "developing-converter",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "simplified-sydney",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "canadian-congress",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vj_annot(df, chains, pairs):\n",
    "    fl_pro_pair_df = pd.DataFrame(df[df['productive']==True].barcode.value_counts())\n",
    "    fl_pro_pair_df = fl_pro_pair_df[fl_pro_pair_df['barcode']>=2]\n",
    "    Result = []\n",
    "    cell_nums = len(set(df['barcode'].tolist()))\n",
    "    Result.append({\n",
    "        'name': 'Cells With Productive V-J Spanning Pair',\n",
    "        'value': fl_pro_pair_df.shape[0],\n",
    "        'total': cell_nums,\n",
    "    })\n",
    "    for p in pairs:\n",
    "        chain1 = p.split('_')[0]\n",
    "        chain2 = p.split('_')[1]\n",
    "        cbs1 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain1)].barcode.tolist())\n",
    "        cbs2 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain2)].barcode.tolist())\n",
    "        paired_cbs = len(cbs1.intersection(cbs2))\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive V-J Spanning ({chain1}, {chain2}) Pair',\n",
    "            'value': paired_cbs,\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "    for c in chains:\n",
    "        Result.append({\n",
    "            'name': f'Cells With {c} Contig',\n",
    "            'value': len(set(df[df['chain']==c].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f'Fraction of cell-associated barcodes with at least one {c} contig annotated as a full or partial V(D)J gene'\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With CDR3-annotated {c} Contig',\n",
    "            'value': len(set(df[(df['chain']==c)&(df['productive']==True)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With V-J Spanning {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f\"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for {c}\"\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "\n",
    "    return Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "amazing-street",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southern-number",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "decimal-forestry",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nutritional-indie",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "japanese-light",
   "metadata": {},
   "outputs": [],
   "source": [
    "#meta = pd.read_csv('/SGRNJ06/randd/USER/cjj/TESTDATA/test_trust/20220530TR_SR/fltest/04.summarize/fltest_assign.csv')\n",
    "#meta =pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/kemai/20220517merTRflilter/mappingmanual/220417014_meta.csv')\n",
    "meta =pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/kemai/20220517merTRflilter/mappingmanual/220417016_meta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "controlled-kuwait",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>sample</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>RNA_snn_res.0.8</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>celltype</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>220417016_AAACATCGAAACATCGGAGCTGAA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>7928</td>\n",
       "      <td>2606</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>GCB</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>220417016_AAACATCGAACAACCACCGAAGTA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>20544</td>\n",
       "      <td>2098</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>PlasmaCells</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>220417016_AAACATCGAACAACCAGCTCGGTA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>30676</td>\n",
       "      <td>2149</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PlasmaCells</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>220417016_AAACATCGAACCGAGAACAGCAGA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>10727</td>\n",
       "      <td>1183</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PlasmaCells</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>220417016_AAACATCGAACCGAGATTCACGCA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>2893</td>\n",
       "      <td>1175</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaiveB</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16350</th>\n",
       "      <td>220417016_TTCACGCATGGCTTCAAAGAGATC</td>\n",
       "      <td>220417016</td>\n",
       "      <td>16833</td>\n",
       "      <td>4074</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>TCells</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16351</th>\n",
       "      <td>220417016_TTCACGCATGGCTTCAAGCACCTC</td>\n",
       "      <td>220417016</td>\n",
       "      <td>3621</td>\n",
       "      <td>1379</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>GCB</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16352</th>\n",
       "      <td>220417016_TTCACGCATGGCTTCAGCTCGGTA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>2161</td>\n",
       "      <td>926</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaiveB</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16353</th>\n",
       "      <td>220417016_TTCACGCATGGCTTCATAGGATGA</td>\n",
       "      <td>220417016</td>\n",
       "      <td>1967</td>\n",
       "      <td>952</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaiveB</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16354</th>\n",
       "      <td>220417016_TTCACGCATTCACGCACCATCCTC</td>\n",
       "      <td>220417016</td>\n",
       "      <td>8726</td>\n",
       "      <td>2751</td>\n",
       "      <td>220417016</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>GCB</td>\n",
       "      <td>T/BCR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16355 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Unnamed: 0  orig.ident  nCount_RNA  \\\n",
       "0      220417016_AAACATCGAAACATCGGAGCTGAA   220417016        7928   \n",
       "1      220417016_AAACATCGAACAACCACCGAAGTA   220417016       20544   \n",
       "2      220417016_AAACATCGAACAACCAGCTCGGTA   220417016       30676   \n",
       "3      220417016_AAACATCGAACCGAGAACAGCAGA   220417016       10727   \n",
       "4      220417016_AAACATCGAACCGAGATTCACGCA   220417016        2893   \n",
       "...                                   ...         ...         ...   \n",
       "16350  220417016_TTCACGCATGGCTTCAAAGAGATC   220417016       16833   \n",
       "16351  220417016_TTCACGCATGGCTTCAAGCACCTC   220417016        3621   \n",
       "16352  220417016_TTCACGCATGGCTTCAGCTCGGTA   220417016        2161   \n",
       "16353  220417016_TTCACGCATGGCTTCATAGGATGA   220417016        1967   \n",
       "16354  220417016_TTCACGCATTCACGCACCATCCTC   220417016        8726   \n",
       "\n",
       "       nFeature_RNA     sample  percent.mt  RNA_snn_res.0.8  seurat_clusters  \\\n",
       "0              2606  220417016           0                7                7   \n",
       "1              2098  220417016           0               10               10   \n",
       "2              2149  220417016           0                4                4   \n",
       "3              1183  220417016           0                4                4   \n",
       "4              1175  220417016           0                2                2   \n",
       "...             ...        ...         ...              ...              ...   \n",
       "16350          4074  220417016           0               12               12   \n",
       "16351          1379  220417016           0                3                3   \n",
       "16352           926  220417016           0                2                2   \n",
       "16353           952  220417016           0                2                2   \n",
       "16354          2751  220417016           0                6                6   \n",
       "\n",
       "          celltype  Class  \n",
       "0              GCB  T/BCR  \n",
       "1      PlasmaCells  T/BCR  \n",
       "2      PlasmaCells  T/BCR  \n",
       "3      PlasmaCells  T/BCR  \n",
       "4           NaiveB  T/BCR  \n",
       "...            ...    ...  \n",
       "16350       TCells  T/BCR  \n",
       "16351          GCB  T/BCR  \n",
       "16352       NaiveB    NaN  \n",
       "16353       NaiveB  T/BCR  \n",
       "16354          GCB  T/BCR  \n",
       "\n",
       "[16355 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "multiple-sperm",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.rename(columns={'Unnamed: 0': 'barcode'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "gentle-siemens",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>sample</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>RNA_snn_res.0.8</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCB</th>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5976</td>\n",
       "      <td>5803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaiveB</th>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>1511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PlasmaCells</th>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2108</td>\n",
       "      <td>2087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ProliferatingBCells</th>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>5679</td>\n",
       "      <td>2095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCells</th>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>913</td>\n",
       "      <td>810</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     barcode  orig.ident  nCount_RNA  nFeature_RNA  sample  \\\n",
       "celltype                                                                     \n",
       "GCB                     5976        5976        5976          5976    5976   \n",
       "NaiveB                  1679        1679        1679          1679    1679   \n",
       "PlasmaCells             2108        2108        2108          2108    2108   \n",
       "ProliferatingBCells     5679        5679        5679          5679    5679   \n",
       "TCells                   913         913         913           913     913   \n",
       "\n",
       "                     percent.mt  RNA_snn_res.0.8  seurat_clusters  Class  \n",
       "celltype                                                                  \n",
       "GCB                        5976             5976             5976   5803  \n",
       "NaiveB                     1679             1679             1679   1511  \n",
       "PlasmaCells                2108             2108             2108   2087  \n",
       "ProliferatingBCells        5679             5679             5679   2095  \n",
       "TCells                      913              913              913    810  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.groupby('celltype').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "secure-camera",
   "metadata": {},
   "outputs": [],
   "source": [
    "tar = {'GCB','NaiveB','PlasmaCells','ProliferatingBCells',}\n",
    "#tar = {'B_cell'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "black-partition",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = meta[meta['celltype'].isin(tar)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "mineral-purchase",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = meta[['barcode']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "official-serial",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta['barcode']= meta['barcode'].apply(lambda x:x.split('_')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "elegant-steel",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv('/SGRNJ06/randd/USER/cjj/TESTDATA/test_trust/barcodelist3.csv',index=False,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ceramic-kansas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACATCGAAACATCGGAGCTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACATCGAACAACCACCGAAGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACATCGAACAACCAGCTCGGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACATCGAACCGAGAACAGCAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACATCGAACCGAGATTCACGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16349</th>\n",
       "      <td>TTCACGCATGGAACAACGAACTTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16351</th>\n",
       "      <td>TTCACGCATGGCTTCAAGCACCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16352</th>\n",
       "      <td>TTCACGCATGGCTTCAGCTCGGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16353</th>\n",
       "      <td>TTCACGCATGGCTTCATAGGATGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16354</th>\n",
       "      <td>TTCACGCATTCACGCACCATCCTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15442 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        barcode\n",
       "0      AAACATCGAAACATCGGAGCTGAA\n",
       "1      AAACATCGAACAACCACCGAAGTA\n",
       "2      AAACATCGAACAACCAGCTCGGTA\n",
       "3      AAACATCGAACCGAGAACAGCAGA\n",
       "4      AAACATCGAACCGAGATTCACGCA\n",
       "...                         ...\n",
       "16349  TTCACGCATGGAACAACGAACTTA\n",
       "16351  TTCACGCATGGCTTCAAGCACCTC\n",
       "16352  TTCACGCATGGCTTCAGCTCGGTA\n",
       "16353  TTCACGCATGGCTTCATAGGATGA\n",
       "16354  TTCACGCATTCACGCACCATCCTC\n",
       "\n",
       "[15442 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stainless-hollow",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
