{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "advised-newsletter",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import re "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "radical-chicken",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_vcf_to_df(vcf_file, cols=('chrom', 'pos', 'alleles'), infos=('VID', 'CID')):\n",
    "    \"\"\"\n",
    "    Read cols and infos into pandas df\n",
    "    \"\"\"\n",
    "    vcf = pysam.VariantFile(vcf_file)\n",
    "    df = pd.DataFrame(columns=[col.capitalize() for col in cols] + infos)\n",
    "    rec_dict = {}\n",
    "    for rec in vcf.fetch():\n",
    "\n",
    "        for col in cols:\n",
    "            rec_dict[col.capitalize()] = getattr(rec, col)\n",
    "            if col == 'alleles':\n",
    "                rec_dict['Alleles'] = '-'.join(rec_dict['Alleles'])\n",
    "\n",
    "        for info in infos:\n",
    "            rec_dict[info] = rec.info[info]\n",
    "\n",
    "        '''\n",
    "        rec_dict['GT'] = [s['GT'] for s in rec.samples.values()][0]\n",
    "        rec_dict['GT'] = [str(item) for item in rec_dict['GT']]\n",
    "        rec_dict['GT'] = '/'.join(rec_dict['GT'])\n",
    "        '''\n",
    "        df_new = pd.DataFrame(rec_dict, index=[0])\n",
    "        df = pd.concat([df, df_new])\n",
    "\n",
    "    vcf.close()\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "baking-scheme",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vcf = parse_vcf_to_df(\"/SGRNJ06/randd/USER/cjj/TESTDATA/git/celescope_test_script/snp/test1/08.filter_snp/test1_filtered.vcf\", infos=[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "respective-meaning",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Alleles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>179218234</td>\n",
       "      <td>G-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>55173993</td>\n",
       "      <td>T-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>55173995</td>\n",
       "      <td>C-CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>55181290</td>\n",
       "      <td>C-G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>55181292</td>\n",
       "      <td>G-T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>55181370</td>\n",
       "      <td>G-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>55181378</td>\n",
       "      <td>C-T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>55191822</td>\n",
       "      <td>T-G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>140753316</td>\n",
       "      <td>A-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7</td>\n",
       "      <td>140753326</td>\n",
       "      <td>T-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7</td>\n",
       "      <td>140753333</td>\n",
       "      <td>T-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7</td>\n",
       "      <td>140753341</td>\n",
       "      <td>A-C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom        Pos Alleles\n",
       "0      3  179218234     G-A\n",
       "1      7   55173993     T-C\n",
       "2      7   55173995    C-CA\n",
       "3      7   55181290     C-G\n",
       "4      7   55181292     G-T\n",
       "5      7   55181370     G-A\n",
       "6      7   55181378     C-T\n",
       "7      7   55191822     T-G\n",
       "8      7  140753316     A-C\n",
       "9      7  140753326     T-A\n",
       "10     7  140753333     T-A\n",
       "11     7  140753341     A-C"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "royal-masters",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ncell = pd.read_csv(\"/SGRNJ06/randd/USER/cjj/TESTDATA/git/celescope_test_script/snp/test1/09.analysis_snp/test1_variant_ncell.csv\")\n",
    "df_vcf = pd.concat([df_vcf, df_ncell], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "combined-guarantee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Pos</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>0/0</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>179218234</td>\n",
       "      <td>G-A</td>\n",
       "      <td>3_179218234</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>55173993</td>\n",
       "      <td>T-C</td>\n",
       "      <td>7_55173993</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>55173995</td>\n",
       "      <td>C-CA</td>\n",
       "      <td>7_55173995</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>55181290</td>\n",
       "      <td>C-G</td>\n",
       "      <td>7_55181290</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>55181292</td>\n",
       "      <td>G-T</td>\n",
       "      <td>7_55181292</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>55181370</td>\n",
       "      <td>G-A</td>\n",
       "      <td>7_55181370</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>55181378</td>\n",
       "      <td>C-T</td>\n",
       "      <td>7_55181378</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>55191822</td>\n",
       "      <td>T-G</td>\n",
       "      <td>7_55191822</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>140753316</td>\n",
       "      <td>A-C</td>\n",
       "      <td>7_140753316</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7</td>\n",
       "      <td>140753326</td>\n",
       "      <td>T-A</td>\n",
       "      <td>7_140753326</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7</td>\n",
       "      <td>140753333</td>\n",
       "      <td>T-A</td>\n",
       "      <td>7_140753333</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7</td>\n",
       "      <td>140753341</td>\n",
       "      <td>A-C</td>\n",
       "      <td>7_140753341</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chrom        Pos Alleles   Unnamed: 0  0/0  0/1  1/1\n",
       "0      3  179218234     G-A  3_179218234    1    0    1\n",
       "1      7   55173993     T-C   7_55173993    3    0    1\n",
       "2      7   55173995    C-CA   7_55173995    2    0    1\n",
       "3      7   55181290     C-G   7_55181290    0    0    2\n",
       "4      7   55181292     G-T   7_55181292    0    0    1\n",
       "5      7   55181370     G-A   7_55181370    0    1    4\n",
       "6      7   55181378     C-T   7_55181378    0    1    4\n",
       "7      7   55191822     T-G   7_55191822    1    0    4\n",
       "8      7  140753316     A-C  7_140753316    1    0    1\n",
       "9      7  140753326     T-A  7_140753326    1    0    1\n",
       "10     7  140753333     T-A  7_140753333    1    0    1\n",
       "11     7  140753341     A-C  7_140753341    1    0    1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "liberal-lindsay",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "tested-centre",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['DP=144;VDB=0;SGB=99.7196;RPB=0.998513;MQB=1;BQB=0.994064;MQ0F=0;AC=2;AN=4;DP4=5,0,139,0;MQ=60;REF_T=10;ALT_T=46;ANN=A', 'missense_variant', 'MODERATE', 'PIK3CA', 'ENSG00000121879', 'transcript', 'ENST00000263967.3', 'protein_coding', '10/21', 'c.1564G>A', 'p.Glu522Lys', '1721/9093', '1564/3207', '522/1068', '', ',A', 'upstream_gene_variant', 'MODIFIER', 'PIK3CA', 'ENSG00000121879', 'transcript', 'ENST00000462255.1', 'retained_intron', '', 'n.-1317G>A', '', '', '', '', '1317', '']\n",
      "PIK3CA\n",
      "1564G>A\n",
      "Glu522Lys\n",
      "=======================================\n",
      "['DP=443;VDB=0.000411984;SGB=1.31889;RPB=0.999946;MQB=1;BQB=0.886634;MQ0F=0;AC=1;AN=10;DP4=403,0,6,0;MQ=60;REF_T=55;ALT_T=10;ANN=C', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '18/28', 'c.2134T>C', 'p.Phe712Leu', '2311/9821', '2134/3633', '712/1210', '', ',C', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '17/26', 'c.1999T>C', 'p.Phe667Leu', '2256/3844', '1999/3276', '667/1091', '', ',C', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '17/27', 'c.1999T>C', 'p.Phe667Leu', '2245/5464', '1999/3498', '667/1165', '', ',C', 'downstream_gene_variant', 'MODIFIER', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000344576.6', 'protein_coding', '', 'c.*3449T>C', '', '', '', '', '2948', '']\n",
      "EGFR\n",
      "2134T>C\n",
      "Phe712Leu\n",
      "1999T>C\n",
      "Phe667Leu\n",
      "1999T>C\n",
      "Phe667Leu\n",
      "*3449T>C\n",
      "=======================================\n",
      "['INDEL;IDV=70;IMF=0.714286;DP=443;VDB=0;SGB=54.8756;MQ0F=0;AC=1;AN=10;DP4=364,0,79,0;MQ=60;REF_T=60;ALT_T=26;ANN=CA', 'frameshift_variant', 'HIGH', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '18/28', 'c.2141dupA', 'p.Ile715fs', '2319/9821', '2142/3633', '714/1210', '', 'INFO_REALIGN_3_PRIME,CA', 'frameshift_variant', 'HIGH', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '17/26', 'c.2006dupA', 'p.Ile670fs', '2264/3844', '2007/3276', '669/1091', '', 'INFO_REALIGN_3_PRIME,CA', 'frameshift_variant', 'HIGH', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '17/27', 'c.2006dupA', 'p.Ile670fs', '2253/5464', '2007/3498', '669/1165', '', 'INFO_REALIGN_3_PRIME,CA', 'downstream_gene_variant', 'MODIFIER', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000344576.6', 'protein_coding', '', 'c.*3451_*3452insA', '', '', '', '', '2951', ';LOF=(EGFR', 'ENSG00000146648', '10', '0.30)']\n",
      "EGFR\n",
      "2141dupA\n",
      "Ile715fs\n",
      "2006dupA\n",
      "Ile670fs\n",
      "2006dupA\n",
      "Ile670fs\n",
      "*3451_*3452insA\n",
      "=======================================\n",
      "['DP=3;VDB=0.02;SGB=-0.00101729;MQ0F=0;AC=4;AN=4;DP4=0,0,2,0;MQ=60;REF_T=10;ALT_T=10;ANN=G', 'splice_region_variant&intron_variant', 'LOW', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '19/27', 'c.2284-3C>G', '', '', '', '', '', ',G', 'splice_region_variant&intron_variant', 'LOW', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '18/25', 'c.2149-3C>G', '', '', '', '', '', ',G', 'splice_region_variant&intron_variant', 'LOW', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '18/26', 'c.2149-3C>G', '', '', '', '', '', ',G', 'non_coding_transcript_exon_variant', 'MODIFIER', 'EGFR-AS1', 'ENSG00000224057', 'transcript', 'ENST00000442411.1', 'antisense', '2/2', 'n.1266G>C', '', '', '', '', '', '']\n",
      "EGFR\n",
      "2284-3C>G\n",
      "2149-3C>G\n",
      "2149-3C>G\n",
      "=======================================\n",
      "['DP=3;VDB=0.02;SGB=-0.00101729;MQ0F=0;AC=4;AN=4;DP4=0,0,2,0;MQ=60;REF_T=10;ALT_T=10;ANN=T', 'splice_acceptor_variant&intron_variant', 'HIGH', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '19/27', 'c.2284-1G>T', '', '', '', '', '', ',T', 'splice_acceptor_variant&intron_variant', 'HIGH', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '18/25', 'c.2149-1G>T', '', '', '', '', '', ',T', 'splice_acceptor_variant&intron_variant', 'HIGH', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '18/26', 'c.2149-1G>T', '', '', '', '', '', ',T', 'non_coding_transcript_exon_variant', 'MODIFIER', 'EGFR-AS1', 'ENSG00000224057', 'transcript', 'ENST00000442411.1', 'antisense', '2/2', 'n.1264C>A', '', '', '', '', '', ';LOF=(EGFR', 'ENSG00000146648', '10', '0.30)']\n",
      "EGFR\n",
      "2284-1G>T\n",
      "2149-1G>T\n",
      "2149-1G>T\n",
      "=======================================\n",
      "['DP=4526;VDB=0;SGB=5268.83;RPB=0.799329;MQB=1;BQB=0.000921768;MQ0F=0;AC=10;AN=10;DP4=177,0,3834,0;MQ=60;REF_T=56;ALT_T=349;ANN=A', 'synonymous_variant', 'LOW', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '20/28', 'c.2361G>A', 'p.Gln787Gln', '2538/9821', '2361/3633', '787/1210', '', ',A', 'synonymous_variant', 'LOW', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '19/26', 'c.2226G>A', 'p.Gln742Gln', '2483/3844', '2226/3276', '742/1091', '', ',A', 'synonymous_variant', 'LOW', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '19/27', 'c.2226G>A', 'p.Gln742Gln', '2472/5464', '2226/3498', '742/1165', '', ',A', 'non_coding_transcript_exon_variant', 'MODIFIER', 'EGFR-AS1', 'ENSG00000224057', 'transcript', 'ENST00000442411.1', 'antisense', '2/2', 'n.1186C>T', '', '', '', '', '', '']\n",
      "EGFR\n",
      "2361G>A\n",
      "Gln787Gln\n",
      "2226G>A\n",
      "Gln742Gln\n",
      "2226G>A\n",
      "Gln742Gln\n",
      "=======================================\n",
      "['DP=4526;VDB=0;SGB=5708.48;RPB=0.607399;MQB=1;BQB=0.336105;MQ0F=0;AC=10;AN=10;DP4=199,0,4186,0;MQ=60;REF_T=57;ALT_T=387;ANN=T', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '20/28', 'c.2369C>T', 'p.Thr790Met', '2546/9821', '2369/3633', '790/1210', '', ',T', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '19/26', 'c.2234C>T', 'p.Thr745Met', '2491/3844', '2234/3276', '745/1091', '', ',T', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '19/27', 'c.2234C>T', 'p.Thr745Met', '2480/5464', '2234/3498', '745/1165', '', ',T', 'non_coding_transcript_exon_variant', 'MODIFIER', 'EGFR-AS1', 'ENSG00000224057', 'transcript', 'ENST00000442411.1', 'antisense', '2/2', 'n.1178G>A', '', '', '', '', '', '']\n",
      "EGFR\n",
      "2369C>T\n",
      "Thr790Met\n",
      "2234C>T\n",
      "Thr745Met\n",
      "2234C>T\n",
      "Thr745Met\n",
      "=======================================\n",
      "['DP=6298;VDB=0;SGB=9092.45;RPB=0.999275;MQB=1;BQB=0.922815;MQ0F=0;AC=10;AN=10;DP4=30,0,6211,0;MQ=60;REF_T=10;ALT_T=617;ANN=G', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000275493.6', 'protein_coding', '21/28', 'c.2573T>G', 'p.Leu858Arg', '2750/9821', '2573/3633', '858/1210', '', ',G', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000455089.5', 'protein_coding', '20/26', 'c.2438T>G', 'p.Leu813Arg', '2695/3844', '2438/3276', '813/1091', '', ',G', 'missense_variant', 'MODERATE', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000454757.6', 'protein_coding', '20/27', 'c.2438T>G', 'p.Leu813Arg', '2684/5464', '2438/3498', '813/1165', '', ',G', 'upstream_gene_variant', 'MODIFIER', 'EGFR-AS1', 'ENSG00000224057', 'transcript', 'ENST00000442411.1', 'antisense', '', 'n.-2888A>C', '', '', '', '', '2888', ',G', 'upstream_gene_variant', 'MODIFIER', 'EGFR', 'ENSG00000146648', 'transcript', 'ENST00000485503.1', 'retained_intron', '', 'n.-989T>G', '', '', '', '', '989', '']\n",
      "EGFR\n",
      "2573T>G\n",
      "Leu858Arg\n",
      "2438T>G\n",
      "Leu813Arg\n",
      "2438T>G\n",
      "Leu813Arg\n",
      "=======================================\n",
      "['DP=3;SGB=-0.000254194;RPB=1;MQB=1;BQB=1;MQ0F=0;AC=1;AN=4;DP4=0,2,0,1;MQ=60;REF_T=10;ALT_T=10;ANN=C', 'missense_variant', 'MODERATE', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000288602.10', 'protein_coding', '15/18', 'c.1819T>G', 'p.Ser607Ala', '1880/2480', '1819/2301', '607/766', '', ',C', 'missense_variant', 'MODERATE', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000496384.6', 'protein_coding', '6/10', 'c.640T>G', 'p.Ser214Ala', '642/8294', '640/1125', '214/374', '', 'WARNING_TRANSCRIPT_NO_START_CODON,C', '3_prime_UTR_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1269T>G', '', '', '', '', '54979', ',C', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000479537.5', 'nonsense_mediated_decay', '2/6', 'n.103T>G', '', '', '', '', '', ',C', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1269T>G', '', '', '', '', '', '']\n",
      "BRAF\n",
      "1819T>G\n",
      "Ser607Ala\n",
      "640T>G\n",
      "Ser214Ala\n",
      "=======================================\n",
      "['DP=3;SGB=-0.000254194;RPB=1;MQB=1;BQB=1;MQ0F=0;AC=2;AN=4;DP4=0,1,0,1;MQ=60;REF_T=10;ALT_T=10;ANN=A', 'synonymous_variant', 'LOW', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000288602.10', 'protein_coding', '15/18', 'c.1809A>T', 'p.Arg603Arg', '1870/2480', '1809/2301', '603/766', '', ',A', 'synonymous_variant', 'LOW', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000496384.6', 'protein_coding', '6/10', 'c.630A>T', 'p.Arg210Arg', '632/8294', '630/1125', '210/374', '', 'WARNING_TRANSCRIPT_NO_START_CODON,A', '3_prime_UTR_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1259A>T', '', '', '', '', '54969', ',A', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000479537.5', 'nonsense_mediated_decay', '2/6', 'n.93A>T', '', '', '', '', '', ',A', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1259A>T', '', '', '', '', '', '']\n",
      "BRAF\n",
      "1809A>T\n",
      "Arg603Arg\n",
      "630A>T\n",
      "Arg210Arg\n",
      "=======================================\n",
      "['DP=3;SGB=-0.000254194;RPB=1;MQB=1;BQB=1;MQ0F=0;AC=1;AN=4;DP4=0,2,0,1;MQ=60;REF_T=10;ALT_T=10;ANN=A', 'missense_variant', 'MODERATE', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000288602.10', 'protein_coding', '15/18', 'c.1802A>T', 'p.Lys601Ile', '1863/2480', '1802/2301', '601/766', '', ',A', 'missense_variant', 'MODERATE', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000496384.6', 'protein_coding', '6/10', 'c.623A>T', 'p.Lys208Ile', '625/8294', '623/1125', '208/374', '', 'WARNING_TRANSCRIPT_NO_START_CODON,A', '3_prime_UTR_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1252A>T', '', '', '', '', '54962', ',A', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000479537.5', 'nonsense_mediated_decay', '2/6', 'n.86A>T', '', '', '', '', '', ',A', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1252A>T', '', '', '', '', '', '']\n",
      "BRAF\n",
      "1802A>T\n",
      "Lys601Ile\n",
      "623A>T\n",
      "Lys208Ile\n",
      "=======================================\n",
      "['DP=3;SGB=-0.000254194;RPB=1;MQB=1;BQB=1;MQ0F=0;AC=1;AN=4;DP4=0,2,0,1;MQ=60;REF_T=10;ALT_T=10;ANN=C', 'synonymous_variant', 'LOW', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000288602.10', 'protein_coding', '15/18', 'c.1794T>G', 'p.Ala598Ala', '1855/2480', '1794/2301', '598/766', '', ',C', 'synonymous_variant', 'LOW', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000496384.6', 'protein_coding', '6/10', 'c.615T>G', 'p.Ala205Ala', '617/8294', '615/1125', '205/374', '', 'WARNING_TRANSCRIPT_NO_START_CODON,C', '3_prime_UTR_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1244T>G', '', '', '', '', '54954', ',C', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000479537.5', 'nonsense_mediated_decay', '2/6', 'n.78T>G', '', '', '', '', '', ',C', 'non_coding_transcript_exon_variant', 'MODIFIER', 'BRAF', 'ENSG00000157764', 'transcript', 'ENST00000497784.1', 'nonsense_mediated_decay', '16/19', 'n.*1244T>G', '', '', '', '', '', '']\n",
      "BRAF\n",
      "1794T>G\n",
      "Ala598Ala\n",
      "615T>G\n",
      "Ala205Ala\n",
      "=======================================\n"
     ]
    }
   ],
   "source": [
    "info_list = []\n",
    "with open(\"/SGRNJ06/randd/USER/cjj/TESTDATA/git/celescope_test_script/snp/test1/08.filter_snp/GRCh99/ann.vcf\") as f:\n",
    "    for line in f.readlines():\n",
    "        if not line.startswith(\"#\"):\n",
    "            info = line.split('\\t')[7]\n",
    "            #info_list.append(info)\n",
    "            #ann_pattern = re.match(r'ANN= \"(\\S+)\"')\n",
    "            anns = info.split(\"|\")\n",
    "            #print(anns)\n",
    "            gene = anns[3]\n",
    "            print(gene)\n",
    "            for ann in anns:\n",
    "                if ann.startswith(\"p.\"):\n",
    "                    protein = ann.strip(\"p.\")\n",
    "                    print(protein)\n",
    "                if ann.startswith(\"c.\"):\n",
    "                    exon = ann.strip(\"c.\")\n",
    "                    print(exon)\n",
    "                \n",
    "            print(\"=======================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "western-orbit",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "centered-drawing",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in info_list:\n",
    "    info = i.split(',')\n",
    "    for j in info:\n",
    "        print(j)\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "suffering-conviction",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "positive-plymouth",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vcf[\"info\"]= info_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "catholic-texture",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "moral-float",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'DP=144;VDB=0;SGB=99.7196;RPB=0.998513;MQB=1;BQB=0.994064;MQ0F=0;AC=2;AN=4;DP4=5,0,139,0;MQ=60;REF_T=10;ALT_T=46;EFF=NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Gaa/Aaa|E522K|1068|PIK3CA|protein_coding|CODING|ENST00000263967|10|A),UPSTREAM(MODIFIER||1317|||PIK3CA|retained_intron|CODING|ENST00000462255||A)'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "accessible-contribution",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'DP=443;VDB=0.000411984;SGB=1.31889;RPB=0.999946;MQB=1;BQB=0.886634;MQ0F=0;AC=1;AN=10;DP4=403,0,6,0;MQ=60;REF_T=55;ALT_T=10;EFF=NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Ttc/Ctc|F712L|1210|EGFR|protein_coding|CODING|ENST00000275493|18|C),NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Ttc/Ctc|F667L|1091|EGFR|protein_coding|CODING|ENST00000455089|17|C),NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|Ttc/Ctc|F667L|1165|EGFR|protein_coding|CODING|ENST00000454757|17|C),DOWNSTREAM(MODIFIER||2948||705|EGFR|protein_coding|CODING|ENST00000344576||C)'"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_list[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "signed-gathering",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cele1.11",
   "language": "python",
   "name": "refactor_trust"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
