{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "nuclear-flexibility",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "previous-rings",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Auto():\n",
    "    \"\"\"\n",
    "    threshold = top {percentile}% cell count / coef\n",
    "    count is usually UMI count.\n",
    "    >>> array = [50] * 100 + [30] * 100 + [10] * 100 + [4] * 100\n",
    "    >>> Auto(array, coef=10).run()\n",
    "    5\n",
    "    >>> Auto(array, percentile=70, coef=3).run()\n",
    "    10\n",
    "    >>> Auto(array, percentile=50, coef=10, expected_cell_num=100).run()\n",
    "    5\n",
    "    >>> Auto([1, 2, 20, 30, 40], expected_cell_num=4, percentile=50, coef=10).run()\n",
    "    2\n",
    "    \"\"\"\n",
    "    def __init__(self, array, percentile=99, coef=3, expected_cell_num=None, **kwargs):\n",
    "        self.array = [x for x in array if x > 0 ]\n",
    "        self.percentile = percentile\n",
    "        self.coef = int(coef)\n",
    "        self.expected_cell_num = expected_cell_num\n",
    "        self.kwargs = kwargs\n",
    "    \n",
    "    def run(self):\n",
    "        array = self.array\n",
    "        if not array:\n",
    "            return 1\n",
    "\n",
    "        if not self.expected_cell_num:\n",
    "            expected_cell_num = len(array)\n",
    "        else:\n",
    "            expected_cell_num = self.expected_cell_num\n",
    "            if expected_cell_num > len(array):\n",
    "                print('Warning: expected_cell_num > len(array)')\n",
    "                expected_cell_num = len(array)\n",
    "                      \n",
    "        sorted_counts = sorted(array, reverse=True)\n",
    "        count_cell_percentile = np.percentile(sorted_counts[:expected_cell_num], self.percentile)\n",
    "        threshold = int(count_cell_percentile / self.coef)\n",
    "\n",
    "        return threshold\n",
    "def target_cell_calling(df_UMI_sum, expected_target_cell_num=3000, target_barcodes=None, weight=6, coef=5, \n",
    "    percentile=85, umi_col='umis'):\n",
    "    \"\"\"\n",
    "    Args:\n",
    "        df_UMI_sum: A dataframe with columns highest umi's contig and UMI.\n",
    "    \n",
    "    Returns:\n",
    "        target_contigs_id: list\n",
    "    >>> df_UMI_sum = pd.DataFrame({\"contig_id\": [\"A\", \"B\", \"C\", \"D\", \"E\"], \"UMI\": [1, 2, 1, 30, 40]})\n",
    "    >>> target_contigs_id = target_cell_calling(df_UMI_sum, expected_target_cell_num=5, percentile=80, coef=5, target_barcodes=[\"A\", \"C\"])\n",
    "    >>> target_contigs_id == {'A_1', 'C_1', 'D_1', 'E_1'}\n",
    "    True\n",
    "    \"\"\"\n",
    "    if target_barcodes != None:\n",
    "        target_barcodes = {i for i in target_barcodes}\n",
    "    umi_threshold = Auto(list(df_UMI_sum[umi_col]), expected_cell_num=expected_target_cell_num, coef=coef, percentile=percentile).run()\n",
    "\n",
    "    # avoid change the original dataframe\n",
    "    df_temp = df_UMI_sum.copy()\n",
    "    if target_barcodes:\n",
    "        df_temp[umi_col] = df_temp.apply(\n",
    "            lambda row:  row[umi_col] * weight if row['barcode'] in target_barcodes else row[umi_col], axis=1)\n",
    "             \n",
    "    target_contigs = set(df_temp.loc[df_temp[umi_col] >= umi_threshold].contig_id)\n",
    "\n",
    "    return target_contigs\n",
    "def get_vj_annot(df, chains, pairs):\n",
    "    fl_pro_pair_df = pd.DataFrame(df[df['productive']==True].barcode.value_counts())\n",
    "    fl_pro_pair_df = fl_pro_pair_df[fl_pro_pair_df['barcode']>=2]\n",
    "    Result = []\n",
    "    cell_nums = len(set(df['barcode'].tolist()))\n",
    "    Result.append({\n",
    "        'name': 'Cells With Productive V-J Spanning Pair',\n",
    "        'value': fl_pro_pair_df.shape[0],\n",
    "        'total': cell_nums,\n",
    "    })\n",
    "    for p in pairs:\n",
    "        chain1 = p.split('_')[0]\n",
    "        chain2 = p.split('_')[1]\n",
    "        cbs1 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain1)].barcode.tolist())\n",
    "        cbs2 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain2)].barcode.tolist())\n",
    "        paired_cbs = len(cbs1.intersection(cbs2))\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive V-J Spanning ({chain1}, {chain2}) Pair',\n",
    "            'value': paired_cbs,\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "    for c in chains:\n",
    "        Result.append({\n",
    "            'name': f'Cells With {c} Contig',\n",
    "            'value': len(set(df[df['chain']==c].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f'Fraction of cell-associated barcodes with at least one {c} contig annotated as a full or partial V(D)J gene'\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With CDR3-annotated {c} Contig',\n",
    "            'value': len(set(df[(df['chain']==c)&(df['cdr3']!=None)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With V-J Spanning {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f\"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for {c}\"\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "\n",
    "    return Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "featured-chart",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/trust_cr_compare/tr_bcr/HA1_0310PBMC_T5_B_3NL/04.summarize/HA1_0310PBMC_T5_B_3NL_b.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "loose-values",
   "metadata": {},
   "outputs": [],
   "source": [
    "        df['productive'] = df['full_length']\n",
    "        contig_set = set(df.contig_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "color-access",
   "metadata": {},
   "outputs": [],
   "source": [
    "        len_dict = dict()\n",
    "\n",
    "        with pysam.FastxFile('/SGRNJ06/randd/USER/cjj/celedev/trust_cr_compare/tr_bcr/HA1_0310PBMC_T5_B_3NL/03.assemble/assemble/HA1_0310PBMC_T5_B_3NL_annotate.fa') as fa:\n",
    "            for read in fa:\n",
    "                len_dict[read.name] = read.comment.split(' ')[0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "funded-missile",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['length'] = df['contig_id'].apply(lambda x: len_dict.get(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "manual-climate",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['length'] = df['contig_id'].apply(len_dict.get)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "informal-focus",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCAGTTCAAACTCACCGCTAACGA</td>\n",
       "      <td>True</td>\n",
       "      <td>CCAGTTCAAACTCACCGCTAACGA_2185</td>\n",
       "      <td>True</td>\n",
       "      <td>482</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV3-10*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ3*02</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CFSTDYTGNVRLF</td>\n",
       "      <td>TGTTTCTCCACAGACTACACTGGTAATGTAAGACTGTTC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CATCAAGTGCGAGTAACGGATTGC</td>\n",
       "      <td>True</td>\n",
       "      <td>CATCAAGTGCGAGTAACGGATTGC_13939</td>\n",
       "      <td>True</td>\n",
       "      <td>545</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-34*01</td>\n",
       "      <td>IGHD3-10*01</td>\n",
       "      <td>IGHJ5*02</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARGKPRITMVRGGRPGWFDPW</td>\n",
       "      <td>TGTGCGAGAGGCAAGCCACGTATTACTATGGTTCGGGGAGGCCGCC...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CATCAAGTGCGAGTAACGGATTGC</td>\n",
       "      <td>True</td>\n",
       "      <td>CATCAAGTGCGAGTAACGGATTGC_33329</td>\n",
       "      <td>True</td>\n",
       "      <td>150</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV1-NL1*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ4*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CQQYYSAPLTF</td>\n",
       "      <td>TGTCAACAGTATTATAGTGCCCCTCTCACTTTC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATTGGCTCAGATCGCAAAGAGATC</td>\n",
       "      <td>True</td>\n",
       "      <td>ATTGGCTCAGATCGCAAAGAGATC_35461</td>\n",
       "      <td>True</td>\n",
       "      <td>150</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-48*03</td>\n",
       "      <td>IGHD3-3*01</td>\n",
       "      <td>IGHJ5*02</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CAREKGFDFSGFDPW</td>\n",
       "      <td>TGTGCGAGAGAGAAAGGGTTCGATTTTTCGGGGTTCGACCCCTGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATTGGCTCAGATCGCAAAGAGATC</td>\n",
       "      <td>True</td>\n",
       "      <td>ATTGGCTCAGATCGCAAAGAGATC_1846</td>\n",
       "      <td>True</td>\n",
       "      <td>510</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV2-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ2*01</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CSSFAGSVLPVIF</td>\n",
       "      <td>TGCTCCTCATTTGCGGGCAGCGTTTTACCTGTCATTTTC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7362</th>\n",
       "      <td>TAGGATGAACCACTGTCATACCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAGGATGAACCACTGTCATACCAA_982</td>\n",
       "      <td>True</td>\n",
       "      <td>573</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ2*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNGPMYTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACGGTCCCATGTACACTTTT</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7363</th>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA</td>\n",
       "      <td>True</td>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA_19492</td>\n",
       "      <td>True</td>\n",
       "      <td>542</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-23*01</td>\n",
       "      <td>IGHD2-8*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CATNRGDYW</td>\n",
       "      <td>TGTGCGACCAACCGAGGGGACTACTGG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7364</th>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA</td>\n",
       "      <td>True</td>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA_6153</td>\n",
       "      <td>True</td>\n",
       "      <td>552</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-20*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ5*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYAGPPITF</td>\n",
       "      <td>TGTCAGCAATATGCTGGGCCACCGATCACCTTC</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7365</th>\n",
       "      <td>GATGAATCGTCTGTCACTAAGGTC</td>\n",
       "      <td>True</td>\n",
       "      <td>GATGAATCGTCTGTCACTAAGGTC_5195</td>\n",
       "      <td>True</td>\n",
       "      <td>295</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-23*01</td>\n",
       "      <td>IGHD6-6*01</td>\n",
       "      <td>IGHJ6*02</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CARELAGRPGRSHYNYYYGMDVW</td>\n",
       "      <td>TGTGCGAGAGAATTAGCAGGTCGTCCCGGGCGGAGTCATTACAATT...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7366</th>\n",
       "      <td>GATGAATCGTCTGTCACTAAGGTC</td>\n",
       "      <td>True</td>\n",
       "      <td>GATGAATCGTCTGTCACTAAGGTC_9046</td>\n",
       "      <td>True</td>\n",
       "      <td>422</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV3-1*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ2*01</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CRAWDDDGVVVF</td>\n",
       "      <td>TGTCGGGCGTGGGACGATGACGGAGTGGTGGTGTTC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7367 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode  is_cell                       contig_id  \\\n",
       "0     CCAGTTCAAACTCACCGCTAACGA     True   CCAGTTCAAACTCACCGCTAACGA_2185   \n",
       "1     CATCAAGTGCGAGTAACGGATTGC     True  CATCAAGTGCGAGTAACGGATTGC_13939   \n",
       "2     CATCAAGTGCGAGTAACGGATTGC     True  CATCAAGTGCGAGTAACGGATTGC_33329   \n",
       "3     ATTGGCTCAGATCGCAAAGAGATC     True  ATTGGCTCAGATCGCAAAGAGATC_35461   \n",
       "4     ATTGGCTCAGATCGCAAAGAGATC     True   ATTGGCTCAGATCGCAAAGAGATC_1846   \n",
       "...                        ...      ...                             ...   \n",
       "7362  TAGGATGAACCACTGTCATACCAA     True    TAGGATGAACCACTGTCATACCAA_982   \n",
       "7363  CAAGACTAAAACATCGGCTCGGTA     True  CAAGACTAAAACATCGGCTCGGTA_19492   \n",
       "7364  CAAGACTAAAACATCGGCTCGGTA     True   CAAGACTAAAACATCGGCTCGGTA_6153   \n",
       "7365  GATGAATCGTCTGTCACTAAGGTC     True   GATGAATCGTCTGTCACTAAGGTC_5195   \n",
       "7366  GATGAATCGTCTGTCACTAAGGTC     True   GATGAATCGTCTGTCACTAAGGTC_9046   \n",
       "\n",
       "      high_confidence length chain        v_gene       d_gene    j_gene  \\\n",
       "0                True    482   IGL   IGLV3-10*01         None  IGLJ3*02   \n",
       "1                True    545   IGH   IGHV4-34*01  IGHD3-10*01  IGHJ5*02   \n",
       "2                True    150   IGK  IGKV1-NL1*01         None  IGKJ4*01   \n",
       "3                True    150   IGH   IGHV3-48*03   IGHD3-3*01  IGHJ5*02   \n",
       "4                True    510   IGL   IGLV2-11*01         None  IGLJ2*01   \n",
       "...               ...    ...   ...           ...          ...       ...   \n",
       "7362             True    573   IGK   IGKV3-11*01         None  IGKJ2*01   \n",
       "7363             True    542   IGH   IGHV3-23*01   IGHD2-8*01  IGHJ4*02   \n",
       "7364             True    552   IGK   IGKV3-20*01         None  IGKJ5*01   \n",
       "7365             True    295   IGH   IGHV3-23*01   IGHD6-6*01  IGHJ6*02   \n",
       "7366             True    422   IGL    IGLV3-1*01         None  IGLJ2*01   \n",
       "\n",
       "     c_gene  full_length  productive                     cdr3  \\\n",
       "0      None        False       False            CFSTDYTGNVRLF   \n",
       "1      IGHM         True        True   CARGKPRITMVRGGRPGWFDPW   \n",
       "2      IGKC        False       False              CQQYYSAPLTF   \n",
       "3      None        False       False          CAREKGFDFSGFDPW   \n",
       "4     IGLC2         True        True            CSSFAGSVLPVIF   \n",
       "...     ...          ...         ...                      ...   \n",
       "7362   IGKC         True        True             CQQRSNGPMYTF   \n",
       "7363   IGHD         True        True                CATNRGDYW   \n",
       "7364   IGKC         True        True              CQQYAGPPITF   \n",
       "7365  IGHG1        False       False  CARELAGRPGRSHYNYYYGMDVW   \n",
       "7366   None        False       False             CRAWDDDGVVVF   \n",
       "\n",
       "                                                cdr3_nt  reads  umis  \\\n",
       "0               TGTTTCTCCACAGACTACACTGGTAATGTAAGACTGTTC    1.0   1.0   \n",
       "1     TGTGCGAGAGGCAAGCCACGTATTACTATGGTTCGGGGAGGCCGCC...    1.0   1.0   \n",
       "2                     TGTCAACAGTATTATAGTGCCCCTCTCACTTTC    2.0   2.0   \n",
       "3         TGTGCGAGAGAGAAAGGGTTCGATTTTTCGGGGTTCGACCCCTGG    1.0   1.0   \n",
       "4               TGCTCCTCATTTGCGGGCAGCGTTTTACCTGTCATTTTC    1.0   1.0   \n",
       "...                                                 ...    ...   ...   \n",
       "7362               TGTCAGCAGCGTAGCAACGGTCCCATGTACACTTTT   12.0  12.0   \n",
       "7363                        TGTGCGACCAACCGAGGGGACTACTGG    2.0   2.0   \n",
       "7364                  TGTCAGCAATATGCTGGGCCACCGATCACCTTC    9.0   9.0   \n",
       "7365  TGTGCGAGAGAATTAGCAGGTCGTCCCGGGCGGAGTCATTACAATT...    1.0   1.0   \n",
       "7366               TGTCGGGCGTGGGACGATGACGGAGTGGTGGTGTTC    1.0   1.0   \n",
       "\n",
       "     raw_clonotype_id raw_consensus_id  \n",
       "0                None             None  \n",
       "1                None             None  \n",
       "2                None             None  \n",
       "3                None             None  \n",
       "4                None             None  \n",
       "...               ...              ...  \n",
       "7362             None             None  \n",
       "7363             None             None  \n",
       "7364             None             None  \n",
       "7365             None             None  \n",
       "7366             None             None  \n",
       "\n",
       "[7367 rows x 18 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "warming-december",
   "metadata": {},
   "outputs": [],
   "source": [
    "l1 = df.length.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unlikely-airplane",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['length'] = df['contig_id'].map(len_dict, na_action='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "likely-yield",
   "metadata": {},
   "outputs": [],
   "source": [
    "l2 = df.length.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "balanced-camping",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(l1)):\n",
    "    if l1[i] != l2[i]:\n",
    "        print(l1[i])\n",
    "        print(l2[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "running-protest",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['length']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "noble-carol",
   "metadata": {},
   "outputs": [],
   "source": [
    "            df_chain_heavy = df[df['chain']=='IGH']\n",
    "            df_chain_light = df[(df['chain']=='IGK') | (df['chain']=='IGL')]\n",
    "            df_chain_heavy = df_chain_heavy.drop_duplicates(['barcode'])\n",
    "            df_chain_light = df_chain_light.drop_duplicates(['barcode'])\n",
    "            df_for_clono = pd.concat([df_chain_heavy, df_chain_light], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "reserved-amsterdam",
   "metadata": {},
   "outputs": [],
   "source": [
    "        trust_report = pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/trust_cr_compare/tr_bcr/HA1_0310PBMC_T5_B_3NL/03.assemble/assemble//HA1_0310PBMC_T5_B_3NL_filter_report.tsv', sep='\\t')\n",
    "        correct_cdr3 = set(df_for_clono.cdr3).intersection(set(trust_report.CDR3aa))\n",
    "        correct_cdr3 = [i for i in correct_cdr3 if i.startswith('C')]\n",
    "        correct_cdr3 = [i for i in correct_cdr3 if len(i)>=5]\n",
    "        correct_cdr3 = [i for i in correct_cdr3 if 'UAG' or 'UAA' or 'UGA' not in i]\n",
    "        df_for_clono = df_for_clono[df_for_clono['cdr3'].isin(correct_cdr3)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "finnish-projection",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATCAAGTGCGAGTAACGGATTGC</td>\n",
       "      <td>True</td>\n",
       "      <td>CATCAAGTGCGAGTAACGGATTGC_13939</td>\n",
       "      <td>True</td>\n",
       "      <td>545</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-34*01</td>\n",
       "      <td>IGHD3-10*01</td>\n",
       "      <td>IGHJ5*02</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARGKPRITMVRGGRPGWFDPW</td>\n",
       "      <td>TGTGCGAGAGGCAAGCCACGTATTACTATGGTTCGGGGAGGCCGCC...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATTGGCTCAGATCGCAAAGAGATC</td>\n",
       "      <td>True</td>\n",
       "      <td>ATTGGCTCAGATCGCAAAGAGATC_35461</td>\n",
       "      <td>True</td>\n",
       "      <td>150</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-48*03</td>\n",
       "      <td>IGHD3-3*01</td>\n",
       "      <td>IGHJ5*02</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CAREKGFDFSGFDPW</td>\n",
       "      <td>TGTGCGAGAGAGAAAGGGTTCGATTTTTCGGGGTTCGACCCCTGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATAGACAAAGAGATCAAGGACAC</td>\n",
       "      <td>True</td>\n",
       "      <td>GATAGACAAAGAGATCAAGGACAC_35465</td>\n",
       "      <td>True</td>\n",
       "      <td>150</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-34*01</td>\n",
       "      <td>IGHD3-22*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CARGWEVVTSFDYW</td>\n",
       "      <td>TGTGCGAGAGGTTGGGAAGTGGTTACCTCTTTTGACTACTGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGCAGGAAAGCACCTCCATACCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>AGCAGGAAAGCACCTCCATACCAA_3757</td>\n",
       "      <td>True</td>\n",
       "      <td>620</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV2-5*01</td>\n",
       "      <td>IGHD3-3*01</td>\n",
       "      <td>IGHJ3*01</td>\n",
       "      <td>IGHG2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAHLDSKKSYTYDFW</td>\n",
       "      <td>TGTGCACACTTGGATTCCAAGAAGTCTTACACTTACGATTTTTGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GACAGTGCCGCTGATCCTAAGGTC</td>\n",
       "      <td>True</td>\n",
       "      <td>GACAGTGCCGCTGATCCTAAGGTC_15483</td>\n",
       "      <td>True</td>\n",
       "      <td>261</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-74*01</td>\n",
       "      <td>IGHD3-22*01</td>\n",
       "      <td>IGHJ5*02</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CARDGYYYDSSGYAGDWFDPW</td>\n",
       "      <td>TGTGCAAGAGATGGGTATTACTATGATAGTAGTGGTTATGCAGGGG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7362</th>\n",
       "      <td>CCAGTTCAGCTAACGATGAAGAGA</td>\n",
       "      <td>True</td>\n",
       "      <td>CCAGTTCAGCTAACGATGAAGAGA_3527</td>\n",
       "      <td>True</td>\n",
       "      <td>571</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-15*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYNAWPRTF</td>\n",
       "      <td>TGTCAGCAGTATAATGCCTGGCCTCGGACGTTC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7363</th>\n",
       "      <td>GCGAGTAAAATGTTGCGTACGCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>GCGAGTAAAATGTTGCGTACGCAA_24183</td>\n",
       "      <td>True</td>\n",
       "      <td>457</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV1-NL1*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ4*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYDNLPLTF</td>\n",
       "      <td>TGTCAACAGTATGATAATCTCCCGCTCACTTTC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7364</th>\n",
       "      <td>TAGGATGAACCACTGTCATACCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAGGATGAACCACTGTCATACCAA_982</td>\n",
       "      <td>True</td>\n",
       "      <td>573</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ2*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNGPMYTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACGGTCCCATGTACACTTTT</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7365</th>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA</td>\n",
       "      <td>True</td>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA_6153</td>\n",
       "      <td>True</td>\n",
       "      <td>552</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-20*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ5*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYAGPPITF</td>\n",
       "      <td>TGTCAGCAATATGCTGGGCCACCGATCACCTTC</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7366</th>\n",
       "      <td>GATGAATCGTCTGTCACTAAGGTC</td>\n",
       "      <td>True</td>\n",
       "      <td>GATGAATCGTCTGTCACTAAGGTC_9046</td>\n",
       "      <td>True</td>\n",
       "      <td>422</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV3-1*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ2*01</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CRAWDDDGVVVF</td>\n",
       "      <td>TGTCGGGCGTGGGACGATGACGGAGTGGTGGTGTTC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7321 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode  is_cell                       contig_id  \\\n",
       "0     CATCAAGTGCGAGTAACGGATTGC     True  CATCAAGTGCGAGTAACGGATTGC_13939   \n",
       "1     ATTGGCTCAGATCGCAAAGAGATC     True  ATTGGCTCAGATCGCAAAGAGATC_35461   \n",
       "2     GATAGACAAAGAGATCAAGGACAC     True  GATAGACAAAGAGATCAAGGACAC_35465   \n",
       "3     AGCAGGAAAGCACCTCCATACCAA     True   AGCAGGAAAGCACCTCCATACCAA_3757   \n",
       "4     GACAGTGCCGCTGATCCTAAGGTC     True  GACAGTGCCGCTGATCCTAAGGTC_15483   \n",
       "...                        ...      ...                             ...   \n",
       "7362  CCAGTTCAGCTAACGATGAAGAGA     True   CCAGTTCAGCTAACGATGAAGAGA_3527   \n",
       "7363  GCGAGTAAAATGTTGCGTACGCAA     True  GCGAGTAAAATGTTGCGTACGCAA_24183   \n",
       "7364  TAGGATGAACCACTGTCATACCAA     True    TAGGATGAACCACTGTCATACCAA_982   \n",
       "7365  CAAGACTAAAACATCGGCTCGGTA     True   CAAGACTAAAACATCGGCTCGGTA_6153   \n",
       "7366  GATGAATCGTCTGTCACTAAGGTC     True   GATGAATCGTCTGTCACTAAGGTC_9046   \n",
       "\n",
       "      high_confidence length chain        v_gene       d_gene    j_gene  \\\n",
       "0                True    545   IGH   IGHV4-34*01  IGHD3-10*01  IGHJ5*02   \n",
       "1                True    150   IGH   IGHV3-48*03   IGHD3-3*01  IGHJ5*02   \n",
       "2                True    150   IGH   IGHV4-34*01  IGHD3-22*01  IGHJ4*02   \n",
       "3                True    620   IGH    IGHV2-5*01   IGHD3-3*01  IGHJ3*01   \n",
       "4                True    261   IGH   IGHV3-74*01  IGHD3-22*01  IGHJ5*02   \n",
       "...               ...    ...   ...           ...          ...       ...   \n",
       "7362             True    571   IGK   IGKV3-15*01         None  IGKJ1*01   \n",
       "7363             True    457   IGK  IGKV1-NL1*01         None  IGKJ4*01   \n",
       "7364             True    573   IGK   IGKV3-11*01         None  IGKJ2*01   \n",
       "7365             True    552   IGK   IGKV3-20*01         None  IGKJ5*01   \n",
       "7366             True    422   IGL    IGLV3-1*01         None  IGLJ2*01   \n",
       "\n",
       "     c_gene  full_length  productive                    cdr3  \\\n",
       "0      IGHM         True        True  CARGKPRITMVRGGRPGWFDPW   \n",
       "1      None        False       False         CAREKGFDFSGFDPW   \n",
       "2      IGHM        False       False          CARGWEVVTSFDYW   \n",
       "3     IGHG2         True        True         CAHLDSKKSYTYDFW   \n",
       "4      IGHD        False       False   CARDGYYYDSSGYAGDWFDPW   \n",
       "...     ...          ...         ...                     ...   \n",
       "7362   IGKC         True        True             CQQYNAWPRTF   \n",
       "7363   IGKC         True        True             CQQYDNLPLTF   \n",
       "7364   IGKC         True        True            CQQRSNGPMYTF   \n",
       "7365   IGKC         True        True             CQQYAGPPITF   \n",
       "7366   None        False       False            CRAWDDDGVVVF   \n",
       "\n",
       "                                                cdr3_nt  reads  umis  \\\n",
       "0     TGTGCGAGAGGCAAGCCACGTATTACTATGGTTCGGGGAGGCCGCC...    1.0   1.0   \n",
       "1         TGTGCGAGAGAGAAAGGGTTCGATTTTTCGGGGTTCGACCCCTGG    1.0   1.0   \n",
       "2            TGTGCGAGAGGTTGGGAAGTGGTTACCTCTTTTGACTACTGG    1.0   1.0   \n",
       "3         TGTGCACACTTGGATTCCAAGAAGTCTTACACTTACGATTTTTGG    1.0   1.0   \n",
       "4     TGTGCAAGAGATGGGTATTACTATGATAGTAGTGGTTATGCAGGGG...   37.0  37.0   \n",
       "...                                                 ...    ...   ...   \n",
       "7362                  TGTCAGCAGTATAATGCCTGGCCTCGGACGTTC    1.0   1.0   \n",
       "7363                  TGTCAACAGTATGATAATCTCCCGCTCACTTTC    2.0   2.0   \n",
       "7364               TGTCAGCAGCGTAGCAACGGTCCCATGTACACTTTT   12.0  12.0   \n",
       "7365                  TGTCAGCAATATGCTGGGCCACCGATCACCTTC    9.0   9.0   \n",
       "7366               TGTCGGGCGTGGGACGATGACGGAGTGGTGGTGTTC    1.0   1.0   \n",
       "\n",
       "     raw_clonotype_id raw_consensus_id  \n",
       "0                None             None  \n",
       "1                None             None  \n",
       "2                None             None  \n",
       "3                None             None  \n",
       "4                None             None  \n",
       "...               ...              ...  \n",
       "7362             None             None  \n",
       "7363             None             None  \n",
       "7364             None             None  \n",
       "7365             None             None  \n",
       "7366             None             None  \n",
       "\n",
       "[7321 rows x 18 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_clono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "written-cornwall",
   "metadata": {},
   "outputs": [],
   "source": [
    "        # 分开过滤\n",
    "        df_chain_heavy = df_for_clono[df_for_clono['chain']=='IGH']\n",
    "        df_chain_light = df_for_clono[(df_for_clono['chain']=='IGK') | (df_for_clono['chain']=='IGL')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "finnish-major",
   "metadata": {},
   "outputs": [],
   "source": [
    "        filtered_congtigs_id = set()\n",
    "        for _df in [df_chain_heavy, df_chain_light]:\n",
    "            target_contigs = target_cell_calling(\n",
    "            _df, \n",
    "            expected_target_cell_num=3000, \n",
    "            target_barcodes=None,\n",
    "            weight = 6.0,\n",
    "            coef = 5\n",
    "            )\n",
    "            filtered_congtigs_id = filtered_congtigs_id | target_contigs       \n",
    "        \n",
    "        df_for_clono = df_for_clono[df_for_clono.contig_id.isin(filtered_congtigs_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fifth-bargain",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GACAGTGCCGCTGATCCTAAGGTC</td>\n",
       "      <td>True</td>\n",
       "      <td>GACAGTGCCGCTGATCCTAAGGTC_15483</td>\n",
       "      <td>True</td>\n",
       "      <td>261</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-74*01</td>\n",
       "      <td>IGHD3-22*01</td>\n",
       "      <td>IGHJ5*02</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CARDGYYYDSSGYAGDWFDPW</td>\n",
       "      <td>TGTGCAAGAGATGGGTATTACTATGATAGTAGTGGTTATGCAGGGG...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CCGAAGTAAGATGTACAGAGTCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>CCGAAGTAAGATGTACAGAGTCAA_11521</td>\n",
       "      <td>True</td>\n",
       "      <td>556</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-39*01</td>\n",
       "      <td>IGHD2-2*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHM</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARREYCSSTDCYDGYW</td>\n",
       "      <td>TGTGCGAGACGGGAATATTGTAGTAGTACCGACTGCTATGATGGTT...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCTCGGTACATACCAATTCACGCA</td>\n",
       "      <td>True</td>\n",
       "      <td>GCTCGGTACATACCAATTCACGCA_16759</td>\n",
       "      <td>True</td>\n",
       "      <td>312</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV4-39*01</td>\n",
       "      <td>IGHD3-22*01</td>\n",
       "      <td>IGHJ6*02</td>\n",
       "      <td>IGHA1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CARHRVVGSFSYGYYYGMDVW</td>\n",
       "      <td>TGTGCGAGGCATCGGGTTGTCGGGAGCTTCTCCTATGGATATTATT...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CATCAAGTGTACGCAAGACTAGTA</td>\n",
       "      <td>True</td>\n",
       "      <td>CATCAAGTGTACGCAAGACTAGTA_8101</td>\n",
       "      <td>True</td>\n",
       "      <td>667</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-30*04</td>\n",
       "      <td>IGHD3-10*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARGGSYFDHW</td>\n",
       "      <td>TGTGCGCGGGGGGGAAGTTATTTTGACCACTGG</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AACTCACCGAGCTGAAACAGATTC</td>\n",
       "      <td>True</td>\n",
       "      <td>AACTCACCGAGCTGAAACAGATTC_2692</td>\n",
       "      <td>True</td>\n",
       "      <td>703</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-15*01</td>\n",
       "      <td>IGHD3-3*01</td>\n",
       "      <td>IGHJ4*02</td>\n",
       "      <td>IGHA2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CTTHPPFYYPSSKKGADYW</td>\n",
       "      <td>TGTACCACACACCCCCCCTTTTATTATCCCTCGTCGAAAAAAGGGG...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7347</th>\n",
       "      <td>CAGATCTGAGCACCTCCAAGGAGC</td>\n",
       "      <td>True</td>\n",
       "      <td>CAGATCTGAGCACCTCCAAGGAGC_14994</td>\n",
       "      <td>True</td>\n",
       "      <td>426</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV3-10*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ3*02</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CFSTDNTGNVRLF</td>\n",
       "      <td>TGTTTCTCCACAGACAACACTGGTAATGTAAGACTGTTC</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7352</th>\n",
       "      <td>CATCAAGTGCGAGTAAACCTCCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>CATCAAGTGCGAGTAAACCTCCAA_6868</td>\n",
       "      <td>True</td>\n",
       "      <td>561</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV1-12*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ5*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQANSFPITF</td>\n",
       "      <td>TGTCAACAGGCTAACAGTTTCCCGATCACCTTC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7359</th>\n",
       "      <td>CTGTAGCCAACGTGATCCTCTATC</td>\n",
       "      <td>True</td>\n",
       "      <td>CTGTAGCCAACGTGATCCTCTATC_17309</td>\n",
       "      <td>True</td>\n",
       "      <td>365</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV3-25*03</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ1*01</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CQSIDSSGTYVF</td>\n",
       "      <td>TGTCAATCAATAGACAGCAGTGGGACTTATGTCTTC</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7364</th>\n",
       "      <td>TAGGATGAACCACTGTCATACCAA</td>\n",
       "      <td>True</td>\n",
       "      <td>TAGGATGAACCACTGTCATACCAA_982</td>\n",
       "      <td>True</td>\n",
       "      <td>573</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-11*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ2*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQRSNGPMYTF</td>\n",
       "      <td>TGTCAGCAGCGTAGCAACGGTCCCATGTACACTTTT</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7365</th>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA</td>\n",
       "      <td>True</td>\n",
       "      <td>CAAGACTAAAACATCGGCTCGGTA_6153</td>\n",
       "      <td>True</td>\n",
       "      <td>552</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV3-20*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ5*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYAGPPITF</td>\n",
       "      <td>TGTCAGCAATATGCTGGGCCACCGATCACCTTC</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2688 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode  is_cell                       contig_id  \\\n",
       "4     GACAGTGCCGCTGATCCTAAGGTC     True  GACAGTGCCGCTGATCCTAAGGTC_15483   \n",
       "8     CCGAAGTAAGATGTACAGAGTCAA     True  CCGAAGTAAGATGTACAGAGTCAA_11521   \n",
       "9     GCTCGGTACATACCAATTCACGCA     True  GCTCGGTACATACCAATTCACGCA_16759   \n",
       "10    CATCAAGTGTACGCAAGACTAGTA     True   CATCAAGTGTACGCAAGACTAGTA_8101   \n",
       "13    AACTCACCGAGCTGAAACAGATTC     True   AACTCACCGAGCTGAAACAGATTC_2692   \n",
       "...                        ...      ...                             ...   \n",
       "7347  CAGATCTGAGCACCTCCAAGGAGC     True  CAGATCTGAGCACCTCCAAGGAGC_14994   \n",
       "7352  CATCAAGTGCGAGTAAACCTCCAA     True   CATCAAGTGCGAGTAAACCTCCAA_6868   \n",
       "7359  CTGTAGCCAACGTGATCCTCTATC     True  CTGTAGCCAACGTGATCCTCTATC_17309   \n",
       "7364  TAGGATGAACCACTGTCATACCAA     True    TAGGATGAACCACTGTCATACCAA_982   \n",
       "7365  CAAGACTAAAACATCGGCTCGGTA     True   CAAGACTAAAACATCGGCTCGGTA_6153   \n",
       "\n",
       "      high_confidence length chain       v_gene       d_gene    j_gene c_gene  \\\n",
       "4                True    261   IGH  IGHV3-74*01  IGHD3-22*01  IGHJ5*02   IGHD   \n",
       "8                True    556   IGH  IGHV4-39*01   IGHD2-2*01  IGHJ4*02   IGHM   \n",
       "9                True    312   IGH  IGHV4-39*01  IGHD3-22*01  IGHJ6*02  IGHA1   \n",
       "10               True    667   IGH  IGHV3-30*04  IGHD3-10*01  IGHJ4*02  IGHG1   \n",
       "13               True    703   IGH  IGHV3-15*01   IGHD3-3*01  IGHJ4*02  IGHA2   \n",
       "...               ...    ...   ...          ...          ...       ...    ...   \n",
       "7347             True    426   IGL  IGLV3-10*01         None  IGLJ3*02   None   \n",
       "7352             True    561   IGK  IGKV1-12*01         None  IGKJ5*01   IGKC   \n",
       "7359             True    365   IGL  IGLV3-25*03         None  IGLJ1*01   None   \n",
       "7364             True    573   IGK  IGKV3-11*01         None  IGKJ2*01   IGKC   \n",
       "7365             True    552   IGK  IGKV3-20*01         None  IGKJ5*01   IGKC   \n",
       "\n",
       "      full_length  productive                   cdr3  \\\n",
       "4           False       False  CARDGYYYDSSGYAGDWFDPW   \n",
       "8            True        True      CARREYCSSTDCYDGYW   \n",
       "9           False       False  CARHRVVGSFSYGYYYGMDVW   \n",
       "10           True        True            CARGGSYFDHW   \n",
       "13           True        True    CTTHPPFYYPSSKKGADYW   \n",
       "...           ...         ...                    ...   \n",
       "7347        False       False          CFSTDNTGNVRLF   \n",
       "7352         True        True            CQQANSFPITF   \n",
       "7359        False       False           CQSIDSSGTYVF   \n",
       "7364         True        True           CQQRSNGPMYTF   \n",
       "7365         True        True            CQQYAGPPITF   \n",
       "\n",
       "                                                cdr3_nt  reads  umis  \\\n",
       "4     TGTGCAAGAGATGGGTATTACTATGATAGTAGTGGTTATGCAGGGG...   37.0  37.0   \n",
       "8     TGTGCGAGACGGGAATATTGTAGTAGTACCGACTGCTATGATGGTT...   24.0  24.0   \n",
       "9     TGTGCGAGGCATCGGGTTGTCGGGAGCTTCTCCTATGGATATTATT...    2.0   2.0   \n",
       "10                    TGTGCGCGGGGGGGAAGTTATTTTGACCACTGG    2.0   2.0   \n",
       "13    TGTACCACACACCCCCCCTTTTATTATCCCTCGTCGAAAAAAGGGG...   16.0  16.0   \n",
       "...                                                 ...    ...   ...   \n",
       "7347            TGTTTCTCCACAGACAACACTGGTAATGTAAGACTGTTC   12.0  12.0   \n",
       "7352                  TGTCAACAGGCTAACAGTTTCCCGATCACCTTC    5.0   5.0   \n",
       "7359               TGTCAATCAATAGACAGCAGTGGGACTTATGTCTTC    7.0   7.0   \n",
       "7364               TGTCAGCAGCGTAGCAACGGTCCCATGTACACTTTT   12.0  12.0   \n",
       "7365                  TGTCAGCAATATGCTGGGCCACCGATCACCTTC    9.0   9.0   \n",
       "\n",
       "     raw_clonotype_id raw_consensus_id  \n",
       "4                None             None  \n",
       "8                None             None  \n",
       "9                None             None  \n",
       "10               None             None  \n",
       "13               None             None  \n",
       "...               ...              ...  \n",
       "7347             None             None  \n",
       "7352             None             None  \n",
       "7359             None             None  \n",
       "7364             None             None  \n",
       "7365             None             None  \n",
       "\n",
       "[2688 rows x 18 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_clono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "neutral-juvenile",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'Cells With Productive V-J Spanning Pair',\n",
       "  'value': 576,\n",
       "  'total': 1793},\n",
       " {'name': 'Cells With Productive V-J Spanning (IGH, IGK, IGH) Pair',\n",
       "  'value': 0,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGH Contig',\n",
       "  'value': 1527,\n",
       "  'total': 1793,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGH contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGH Contig',\n",
       "  'value': 1527,\n",
       "  'total': 1793},\n",
       " {'name': 'Cells With V-J Spanning IGH Contig',\n",
       "  'value': 1157,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGH\"},\n",
       " {'name': 'Cells With Productive IGH Contig',\n",
       "  'value': 1157,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGK Contig',\n",
       "  'value': 670,\n",
       "  'total': 1793,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGK contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGK Contig', 'value': 670, 'total': 1793},\n",
       " {'name': 'Cells With V-J Spanning IGK Contig',\n",
       "  'value': 540,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGK\"},\n",
       " {'name': 'Cells With Productive IGK Contig',\n",
       "  'value': 540,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGL Contig',\n",
       "  'value': 491,\n",
       "  'total': 1793,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGL contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGL Contig', 'value': 491, 'total': 1793},\n",
       " {'name': 'Cells With V-J Spanning IGL Contig',\n",
       "  'value': 408,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGL\"},\n",
       " {'name': 'Cells With Productive IGL Contig',\n",
       "  'value': 408,\n",
       "  'total': 1793,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"}]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_vj_annot(df_for_clono, ['IGH', 'IGK','IGL'], ['IGH_IGK, IGH_IGL'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "personalized-courage",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UMI_sum = df_for_clono.groupby(['barcode'], as_index=False).agg({\"umis\": \"sum\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "integrated-resident",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UMI_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "honey-celebration",
   "metadata": {},
   "outputs": [],
   "source": [
    "            target_barcodes = target_cell_calling(\n",
    "            df_UMI_sum, \n",
    "            expected_target_cell_num=3000, \n",
    "            target_barcodes=None,\n",
    "            weight = 6.0,\n",
    "            coef = 5\n",
    "            )   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "solar-approval",
   "metadata": {},
   "outputs": [],
   "source": [
    "        \n",
    "        df_for_clono = df_for_clono[df_for_clono.barcode.isin(target_barcodes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exclusive-fence",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_for_clono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nervous-block",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "therapeutic-wrong",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moving-verse",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_for_clono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suitable-surge",
   "metadata": {},
   "outputs": [],
   "source": [
    "        df_for_clono_pro = df_for_clono[df_for_clono['productive']==True]\n",
    "        cell_barcodes = set(df_for_clono_pro['barcode'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "outdoor-growth",
   "metadata": {},
   "outputs": [],
   "source": [
    "        df_filter = df[df.barcode.isin(cell_barcodes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "valued-position",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_vj_annot(df_filter, ['TRA', 'TRB'], ['TRA_TRB'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ongoing-thousand",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "brief-header",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vj_annot(df, chains, pairs):\n",
    "    fl_pro_pair_df = pd.DataFrame(df[df['productive']==True].barcode.value_counts())\n",
    "    fl_pro_pair_df = fl_pro_pair_df[fl_pro_pair_df['barcode']>=2]\n",
    "    Result = []\n",
    "    cell_nums = len(set(df['barcode'].tolist()))\n",
    "    Result.append({\n",
    "        'name': 'Cells With Productive V-J Spanning Pair',\n",
    "        'value': fl_pro_pair_df.shape[0],\n",
    "        'total': cell_nums,\n",
    "    })\n",
    "    for p in pairs:\n",
    "        chain1 = p.split('_')[0]\n",
    "        chain2 = p.split('_')[1]\n",
    "        cbs1 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain1)].barcode.tolist())\n",
    "        cbs2 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain2)].barcode.tolist())\n",
    "        paired_cbs = len(cbs1.intersection(cbs2))\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive V-J Spanning ({chain1}, {chain2}) Pair',\n",
    "            'value': paired_cbs,\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "    for c in chains:\n",
    "        Result.append({\n",
    "            'name': f'Cells With {c} Contig',\n",
    "            'value': len(set(df[df['chain']==c].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f'Fraction of cell-associated barcodes with at least one {c} contig annotated as a full or partial V(D)J gene'\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With CDR3-annotated {c} Contig',\n",
    "            'value': len(set(df[(df['chain']==c)&(df['cdr3']!=None)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With V-J Spanning {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f\"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for {c}\"\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "\n",
    "    return Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "anticipated-arrival",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vj_annot(df, chains, pairs):\n",
    "    fl_pro_pair_df = pd.DataFrame(df.barcode.value_counts())\n",
    "    fl_pro_pair_df_one_chain_bc = set(fl_pro_pair_df[fl_pro_pair_df.barcode==1].index)\n",
    "    fl_pro_pair_df = df[df.barcode.isin(fl_pro_pair_df_one_chain_bc)]\n",
    "    count = fl_pro_pair_df[fl_pro_pair_df['productive']==True].shape[0]\n",
    "    \n",
    "    fl_pro_pair_df = pd.DataFrame(df[df['productive']==True].barcode.value_counts())\n",
    "    count += fl_pro_pair_df[fl_pro_pair_df['barcode']>=2].shape[0]\n",
    "    \n",
    "    Result = []\n",
    "    cell_nums = len(set(df['barcode'].tolist()))\n",
    "    Result.append({\n",
    "        'name': 'Cells With Productive V-J Spanning Pair',\n",
    "        'value': count,\n",
    "        'total': cell_nums,\n",
    "    })\n",
    "    for p in pairs:\n",
    "        chain1 = p.split('_')[0]\n",
    "        chain2 = p.split('_')[1]\n",
    "        cbs1 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain1)].barcode.tolist())\n",
    "        cbs2 = set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==chain2)].barcode.tolist())\n",
    "        paired_cbs = len(cbs1.intersection(cbs2))\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive V-J Spanning ({chain1}, {chain2}) Pair',\n",
    "            'value': paired_cbs,\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "    for c in chains:\n",
    "        Result.append({\n",
    "            'name': f'Cells With {c} Contig',\n",
    "            'value': len(set(df[df['chain']==c].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f'Fraction of cell-associated barcodes with at least one {c} contig annotated as a full or partial V(D)J gene'\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With CDR3-annotated {c} Contig',\n",
    "            'value': len(set(df[(df['chain']==c)&(df['cdr3']!=None)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With V-J Spanning {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': f\"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for {c}\"\n",
    "        })\n",
    "        Result.append({\n",
    "            'name': f'Cells With Productive {c} Contig',\n",
    "            'value': len(set(df[(df['full_length']==True)&(df['productive']==True)&(df['chain']==c)].barcode.tolist())),\n",
    "            'total': cell_nums,\n",
    "            'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"\n",
    "        })\n",
    "\n",
    "    return Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "hidden-madrid",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_filter= pd.read_csv('/SGRNJ06/randd/USER/cjj/celedev/trust_cr_compare/tr_bcr/HA1_0310PBMC_T5_B_3NL/04.summarize/HA1_0310PBMC_T5_B_3NL_filtered_contig.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "moral-barrel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'Cells With Productive V-J Spanning Pair',\n",
       "  'value': 1000,\n",
       "  'total': 1529},\n",
       " {'name': 'Cells With Productive V-J Spanning (IGH, IGK) Pair',\n",
       "  'value': 499,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With Productive V-J Spanning (IGH, IGL) Pair',\n",
       "  'value': 395,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with one productive contig for each chain of the receptor pair.A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGH Contig',\n",
       "  'value': 1444,\n",
       "  'total': 1529,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGH contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGH Contig',\n",
       "  'value': 1444,\n",
       "  'total': 1529},\n",
       " {'name': 'Cells With V-J Spanning IGH Contig',\n",
       "  'value': 1218,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGH\"},\n",
       " {'name': 'Cells With Productive IGH Contig',\n",
       "  'value': 1218,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGK Contig',\n",
       "  'value': 845,\n",
       "  'total': 1529,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGK contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGK Contig', 'value': 845, 'total': 1529},\n",
       " {'name': 'Cells With V-J Spanning IGK Contig',\n",
       "  'value': 678,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGK\"},\n",
       " {'name': 'Cells With Productive IGK Contig',\n",
       "  'value': 678,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"},\n",
       " {'name': 'Cells With IGL Contig',\n",
       "  'value': 663,\n",
       "  'total': 1529,\n",
       "  'help_info': 'Fraction of cell-associated barcodes with at least one IGL contig annotated as a full or partial V(D)J gene'},\n",
       " {'name': 'Cells With CDR3-annotated IGL Contig', 'value': 663, 'total': 1529},\n",
       " {'name': 'Cells With V-J Spanning IGL Contig',\n",
       "  'value': 527,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with at least one contig spanning the 5' end of the V region to the 3' end of the J region for IGL\"},\n",
       " {'name': 'Cells With Productive IGL Contig',\n",
       "  'value': 527,\n",
       "  'total': 1529,\n",
       "  'help_info': \"Fraction of cell-associated barcodes with productive IGL chain. A productive contig satisfies the following conditions: the contig annotations span the 5' end of the V region to the 3' end of the J region of the chain, a start codon was found in the expected part of the V sequence, an in-frame CDR3 amino acid motif was found, and no stop codons were found in the aligned V-J region\"}]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_vj_annot(df_filter, ['IGH','IGK','IGL'], ['IGH_IGK', 'IGH_IGL'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "together-lindsay",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_filter_filter = pd.DataFrame(df_filter.barcode.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "fabulous-anthony",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTACGCAAAAGAGATCCGCTGATC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTACGCAAACAGATTCGTGTTCTA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCCATGCGCTAACGAGTCTGTCA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCACCTCATAGCGACCCTAATCC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGATGCCTAACAATGGAA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTGGCATAATTGGCTCAATGTTGC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAGATTCAGCACCTCAATGTTGC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGACACACTATCAGCAACCTCCAA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTAACGACAACCACAGTCGTAGA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGAGACCATCCTCCGCTGATC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          barcode\n",
       "GTACGCAAAAGAGATCCGCTGATC        1\n",
       "GTACGCAAACAGATTCGTGTTCTA        1\n",
       "AGCCATGCGCTAACGAGTCTGTCA        1\n",
       "AGCACCTCATAGCGACCCTAATCC        1\n",
       "AAACATCGATGCCTAACAATGGAA        1\n",
       "...                           ...\n",
       "CTGGCATAATTGGCTCAATGTTGC        1\n",
       "ACAGATTCAGCACCTCAATGTTGC        1\n",
       "CGACACACTATCAGCAACCTCCAA        1\n",
       "GCTAACGACAACCACAGTCGTAGA        1\n",
       "CCGTGAGACCATCCTCCGCTGATC        1\n",
       "\n",
       "[106 rows x 1 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filter_filter[df_filter_filter.barcode==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "useful-voluntary",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>clonotype_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>CCGTGAGACCATCCTCCGCTGATC</td>\n",
       "      <td>True</td>\n",
       "      <td>CCGTGAGACCATCCTCCGCTGATC_3090</td>\n",
       "      <td>True</td>\n",
       "      <td>573</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV1-39*01</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1*01</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQSYSTPWTF</td>\n",
       "      <td>TGTCAACAGAGTTACAGTACCCCGTGGACGTTC</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>clonotype78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       barcode  is_cell                      contig_id  \\\n",
       "1000  CCGTGAGACCATCCTCCGCTGATC     True  CCGTGAGACCATCCTCCGCTGATC_3090   \n",
       "\n",
       "      high_confidence  length chain       v_gene d_gene    j_gene c_gene  \\\n",
       "1000             True     573   IGK  IGKV1-39*01   None  IGKJ1*01   IGKC   \n",
       "\n",
       "      full_length  productive         cdr3                            cdr3_nt  \\\n",
       "1000         True        True  CQQSYSTPWTF  TGTCAACAGAGTTACAGTACCCCGTGGACGTTC   \n",
       "\n",
       "      reads  umis clonotype_id  \n",
       "1000    5.0   5.0  clonotype78  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filter[df_filter['barcode']=='CCGTGAGACCATCCTCCGCTGATC']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "quantitative-stadium",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "greatest-crisis",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "sustained-party",
   "metadata": {},
   "outputs": [],
   "source": [
    "ten = pd.read_csv('/SGRNJ03/randd/RD20073101_SCOPEv2_TCR/20220322_1/HA1_0310PBMC_T5_B_3NL/03.assemble/HA1_0310PBMC_T5_B_3NL/outs/filtered_contig_annotations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "regular-moldova",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGAAACCAT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCAT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>801</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV2-8</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CSSYAGGHNVGVF</td>\n",
       "      <td>TGCAGCTCATATGCAGGCGGCCACAATGTGGGGGTATTC</td>\n",
       "      <td>66421</td>\n",
       "      <td>3365</td>\n",
       "      <td>clonotype9</td>\n",
       "      <td>clonotype9_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGAAACCAT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCAT-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>576</td>\n",
       "      <td>Multi</td>\n",
       "      <td>IGHV3-30</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ1</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CVKWHQAF</td>\n",
       "      <td>TGTGTGAAATGGCACCAGGCGTTC</td>\n",
       "      <td>4602</td>\n",
       "      <td>186</td>\n",
       "      <td>clonotype9</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGAAACCTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>647</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV6-57</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ3</td>\n",
       "      <td>IGLC2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQSYDNREEDSIWMF</td>\n",
       "      <td>TGTCAGTCTTATGATAATAGAGAGGAAGACAGTATTTGGATGTTC</td>\n",
       "      <td>36023</td>\n",
       "      <td>3448</td>\n",
       "      <td>clonotype21</td>\n",
       "      <td>clonotype21_consensus_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGAAACCTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>599</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-48</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHD</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARATIETFYYFDSW</td>\n",
       "      <td>TGTGCGAGGGCGACTATCGAAACTTTTTACTATTTTGACTCCTGG</td>\n",
       "      <td>33484</td>\n",
       "      <td>771</td>\n",
       "      <td>clonotype21</td>\n",
       "      <td>clonotype21_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGAGAAACCTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGAGAAACCTA-1_contig_3</td>\n",
       "      <td>True</td>\n",
       "      <td>555</td>\n",
       "      <td>IGL</td>\n",
       "      <td>IGLV2-8</td>\n",
       "      <td>None</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CSSYAGSNKVF</td>\n",
       "      <td>TGCAGCTCATATGCAGGCAGCAACAAAGTATTC</td>\n",
       "      <td>41</td>\n",
       "      <td>9</td>\n",
       "      <td>clonotype21</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3325</th>\n",
       "      <td>ACATCAGCATAGTAAG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>ACATCAGCATAGTAAG-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>527</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV1D-13</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ2</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYYGTPRTF</td>\n",
       "      <td>TGTCAACAGTATTATGGTACCCCTCGAACTTTT</td>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype930</td>\n",
       "      <td>clonotype930_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3326</th>\n",
       "      <td>ACATCAGCATAGTAAG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>ACATCAGCATAGTAAG-1_contig_2</td>\n",
       "      <td>True</td>\n",
       "      <td>638</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-33</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>IGHA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CAKAQDILTGLLHFGMDVW</td>\n",
       "      <td>TGTGCGAAAGCGCAGGATATTTTGACTGGATTACTCCACTTCGGTA...</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>clonotype930</td>\n",
       "      <td>clonotype930_consensus_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3327</th>\n",
       "      <td>ACATCAGCATAGTAAG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>ACATCAGCATAGTAAG-1_contig_3</td>\n",
       "      <td>True</td>\n",
       "      <td>374</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ2</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CQQYYGTPRTF</td>\n",
       "      <td>TGTCAACAGTATTATGGTACCCCTCGAACTTTT</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>clonotype930</td>\n",
       "      <td>clonotype930_consensus_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3328</th>\n",
       "      <td>ACATCAGCATAGTAAG-1</td>\n",
       "      <td>True</td>\n",
       "      <td>ACATCAGCATAGTAAG-1_contig_4</td>\n",
       "      <td>True</td>\n",
       "      <td>576</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHJ6</td>\n",
       "      <td>IGHA1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>CAKAQDILTGLLHFGMDVW</td>\n",
       "      <td>TGTGCGAAAGCGCAGGATATTTTGACTGGATTACTCCACTTCGGTA...</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>clonotype930</td>\n",
       "      <td>clonotype930_consensus_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3329</th>\n",
       "      <td>ACCGTAATCTTACCTA-1</td>\n",
       "      <td>True</td>\n",
       "      <td>ACCGTAATCTTACCTA-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>656</td>\n",
       "      <td>IGH</td>\n",
       "      <td>IGHV5-51</td>\n",
       "      <td>None</td>\n",
       "      <td>IGHJ2</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CARHTTAQMKQWPYGDLW</td>\n",
       "      <td>TGTGCGAGACATACCACAGCGCAAATGAAGCAGTGGCCATACGGCG...</td>\n",
       "      <td>78</td>\n",
       "      <td>2</td>\n",
       "      <td>clonotype931</td>\n",
       "      <td>clonotype931_consensus_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3330 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 barcode  is_cell                    contig_id  \\\n",
       "0     AAACCTGAGAAACCAT-1     True  AAACCTGAGAAACCAT-1_contig_1   \n",
       "1     AAACCTGAGAAACCAT-1     True  AAACCTGAGAAACCAT-1_contig_2   \n",
       "2     AAACCTGAGAAACCTA-1     True  AAACCTGAGAAACCTA-1_contig_1   \n",
       "3     AAACCTGAGAAACCTA-1     True  AAACCTGAGAAACCTA-1_contig_2   \n",
       "4     AAACCTGAGAAACCTA-1     True  AAACCTGAGAAACCTA-1_contig_3   \n",
       "...                  ...      ...                          ...   \n",
       "3325  ACATCAGCATAGTAAG-1     True  ACATCAGCATAGTAAG-1_contig_1   \n",
       "3326  ACATCAGCATAGTAAG-1     True  ACATCAGCATAGTAAG-1_contig_2   \n",
       "3327  ACATCAGCATAGTAAG-1     True  ACATCAGCATAGTAAG-1_contig_3   \n",
       "3328  ACATCAGCATAGTAAG-1     True  ACATCAGCATAGTAAG-1_contig_4   \n",
       "3329  ACCGTAATCTTACCTA-1     True  ACCGTAATCTTACCTA-1_contig_1   \n",
       "\n",
       "      high_confidence  length  chain     v_gene d_gene j_gene c_gene  \\\n",
       "0                True     801    IGL    IGLV2-8   None  IGLJ2  IGLC2   \n",
       "1                True     576  Multi   IGHV3-30   None  IGKJ1   IGHD   \n",
       "2                True     647    IGL   IGLV6-57   None  IGLJ3  IGLC2   \n",
       "3                True     599    IGH   IGHV3-48   None  IGHJ4   IGHD   \n",
       "4                True     555    IGL    IGLV2-8   None  IGLJ2   None   \n",
       "...               ...     ...    ...        ...    ...    ...    ...   \n",
       "3325             True     527    IGK  IGKV1D-13   None  IGKJ2   IGKC   \n",
       "3326             True     638    IGH   IGHV3-33   None  IGHJ6  IGHA1   \n",
       "3327             True     374    IGK    IGKV4-1   None  IGKJ2   IGKC   \n",
       "3328             True     576    IGH   IGHV3-23   None  IGHJ6  IGHA1   \n",
       "3329             True     656    IGH   IGHV5-51   None  IGHJ2  IGHG1   \n",
       "\n",
       "      full_length  productive                 cdr3  \\\n",
       "0            True        True        CSSYAGGHNVGVF   \n",
       "1            True       False             CVKWHQAF   \n",
       "2            True        True      CQSYDNREEDSIWMF   \n",
       "3            True        True      CARATIETFYYFDSW   \n",
       "4            True       False          CSSYAGSNKVF   \n",
       "...           ...         ...                  ...   \n",
       "3325         True        True          CQQYYGTPRTF   \n",
       "3326         True        True  CAKAQDILTGLLHFGMDVW   \n",
       "3327         True       False          CQQYYGTPRTF   \n",
       "3328         True       False  CAKAQDILTGLLHFGMDVW   \n",
       "3329         True        True   CARHTTAQMKQWPYGDLW   \n",
       "\n",
       "                                                cdr3_nt  reads  umis  \\\n",
       "0               TGCAGCTCATATGCAGGCGGCCACAATGTGGGGGTATTC  66421  3365   \n",
       "1                              TGTGTGAAATGGCACCAGGCGTTC   4602   186   \n",
       "2         TGTCAGTCTTATGATAATAGAGAGGAAGACAGTATTTGGATGTTC  36023  3448   \n",
       "3         TGTGCGAGGGCGACTATCGAAACTTTTTACTATTTTGACTCCTGG  33484   771   \n",
       "4                     TGCAGCTCATATGCAGGCAGCAACAAAGTATTC     41     9   \n",
       "...                                                 ...    ...   ...   \n",
       "3325                  TGTCAACAGTATTATGGTACCCCTCGAACTTTT     26     3   \n",
       "3326  TGTGCGAAAGCGCAGGATATTTTGACTGGATTACTCCACTTCGGTA...     53     1   \n",
       "3327                  TGTCAACAGTATTATGGTACCCCTCGAACTTTT     12     2   \n",
       "3328  TGTGCGAAAGCGCAGGATATTTTGACTGGATTACTCCACTTCGGTA...     20     3   \n",
       "3329  TGTGCGAGACATACCACAGCGCAAATGAAGCAGTGGCCATACGGCG...     78     2   \n",
       "\n",
       "     raw_clonotype_id          raw_consensus_id  \n",
       "0          clonotype9    clonotype9_consensus_1  \n",
       "1          clonotype9                      None  \n",
       "2         clonotype21   clonotype21_consensus_2  \n",
       "3         clonotype21   clonotype21_consensus_1  \n",
       "4         clonotype21                      None  \n",
       "...               ...                       ...  \n",
       "3325     clonotype930  clonotype930_consensus_1  \n",
       "3326     clonotype930  clonotype930_consensus_2  \n",
       "3327     clonotype930  clonotype930_consensus_1  \n",
       "3328     clonotype930  clonotype930_consensus_2  \n",
       "3329     clonotype931  clonotype931_consensus_1  \n",
       "\n",
       "[3330 rows x 18 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ten"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "upset-magnitude",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1124"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(ten.barcode))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "electronic-addition",
   "metadata": {},
   "outputs": [],
   "source": [
    "flten = pd.DataFrame(ten[ten['productive']==True].barcode.value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "afraid-vacation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACAGAGGT-1</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGATGTGG-1</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGTAGCAAGCCCAC-1</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGCACTA-1</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCGCGAGTACGTTC-1</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGTCCCGACTT-1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACACGTGTTAGGGTG-1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACACGTGTTCGTTGA-1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCATCGATG-1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCACGCGGT-1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1124 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    barcode\n",
       "AAACCTGCACAGAGGT-1        8\n",
       "AAACCTGGTGATGTGG-1        7\n",
       "AAAGTAGCAAGCCCAC-1        7\n",
       "AAACCTGAGGGCACTA-1        7\n",
       "AACCGCGAGTACGTTC-1        7\n",
       "...                     ...\n",
       "AAAGATGTCCCGACTT-1        1\n",
       "AACACGTGTTAGGGTG-1        1\n",
       "AACACGTGTTCGTTGA-1        1\n",
       "AAACCTGTCATCGATG-1        1\n",
       "AAACCTGTCACGCGGT-1        1\n",
       "\n",
       "[1124 rows x 1 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flten"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "functional-essex",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1087"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flten[flten['barcode']>=2].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "general-delivery",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>high_confidence</th>\n",
       "      <th>length</th>\n",
       "      <th>chain</th>\n",
       "      <th>v_gene</th>\n",
       "      <th>d_gene</th>\n",
       "      <th>j_gene</th>\n",
       "      <th>c_gene</th>\n",
       "      <th>full_length</th>\n",
       "      <th>productive</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>cdr3_nt</th>\n",
       "      <th>reads</th>\n",
       "      <th>umis</th>\n",
       "      <th>raw_clonotype_id</th>\n",
       "      <th>raw_consensus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1049</th>\n",
       "      <td>AAACCTGTCACGCGGT-1</td>\n",
       "      <td>True</td>\n",
       "      <td>AAACCTGTCACGCGGT-1_contig_1</td>\n",
       "      <td>True</td>\n",
       "      <td>743</td>\n",
       "      <td>IGK</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>None</td>\n",
       "      <td>IGKJ2</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CQQYYTTPNTF</td>\n",
       "      <td>TGTCAACAATATTATACTACTCCTAACACTTTT</td>\n",
       "      <td>1137</td>\n",
       "      <td>4</td>\n",
       "      <td>clonotype303</td>\n",
       "      <td>clonotype303_consensus_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 barcode  is_cell                    contig_id  \\\n",
       "1049  AAACCTGTCACGCGGT-1     True  AAACCTGTCACGCGGT-1_contig_1   \n",
       "\n",
       "      high_confidence  length chain   v_gene d_gene j_gene c_gene  \\\n",
       "1049             True     743   IGK  IGKV4-1   None  IGKJ2   IGKC   \n",
       "\n",
       "      full_length  productive         cdr3                            cdr3_nt  \\\n",
       "1049         True        True  CQQYYTTPNTF  TGTCAACAATATTATACTACTCCTAACACTTTT   \n",
       "\n",
       "      reads  umis raw_clonotype_id          raw_consensus_id  \n",
       "1049   1137     4     clonotype303  clonotype303_consensus_1  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ten[ten['barcode']=='AAACCTGTCACGCGGT-1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interested-warrior",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
